\documentclass[12pt,]{article}
\usepackage{lmodern}
\usepackage{setspace}
\setstretch{1.5}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1in,letterpaper]{geometry}
\usepackage{hyperref}
\PassOptionsToPackage{usenames,dvipsnames}{color} % color is loaded by hyperref
\hypersetup{unicode=true,
            pdftitle={gemino: An efficient, flexible molecular evolution and sequencing simulator},
            colorlinks=true,
            linkcolor=Maroon,
            citecolor=Blue,
            urlcolor=Blue,
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\newcommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

  \title{gemino: An efficient, flexible molecular evolution and sequencing
simulator}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
    \author{}
    \preauthor{}\postauthor{}
    \date{}
    \predate{}\postdate{}
  

\begin{document}
\maketitle

\hypertarget{abstract}{%
\subsection{Abstract}\label{abstract}}

High-throughput sequencing (HTS) is central to the study of population
genomics. Choices in sampling design for sequencing projects can include
sequencing method (e.g., restriction-site associated DNA sequencing
{[}RADseq{]} versus whole genome sequencing {[}WGS{]}), depth of
coverage, number of individuals to sample, and selecting the appropriate
restriction enzyme(s) (for RADseq). These choices are most often
informed by previous work on highly diverged species, which ignores
species- and population-specific genomic characteristics, demographies,
and evolutionary histories. Simulating sequencing based on available
genomic data better informs sampling strategies. However, current
methods provide only rudimentary ways to simulate population structure
and variation in coverage among sites. Here I present the R package
\texttt{gemino} that efficiently (i) reads and simulates reference
genomes; (ii) generates variants using summary statistics, phylogenies,
Variant Call Format (VCF) files, and coalescent simulations---the latter
of which can include selection, recombination, and demographic
fluctuations; (iii) simulates sequencing error, mapping qualities,
restriction-enzyme digestion, and variance in coverage among sites; and
(iv) writes outputs to standard file formats. \texttt{gemino} can
simulate single or paired-ended reads for WGS, RADseq, or
genotyping-by-sequencing on the Illumina platform. Although defaults are
provided for most functions, relatively few aspects of \texttt{gemino}
are hard-coded, providing the user flexibility for their simulations.
\texttt{gemino} can be extended to simulate different sequencing
technologies, such as Pacific BioSciences, or Oxford Nanopore
Technologies. Most functions are written in C++ to improve performance,
and I employed OpenMP to allow for parallel processing. The stable
version of \texttt{gemino} is available on CRAN
(\url{https://CRAN.R-project.org/package=gemino}), and the development
version is on GitHub (\url{https://github.com/lucasnell/gemino}).

\textbf{Keywords:} sequencing simulator, population genomics,
high-throughput sequencing, Illumina, Pool-seq, RADseq

\hypertarget{introduction}{%
\subsection{Introduction}\label{introduction}}

High-throughput sequencing (HTS) is a cost-effective approach to
generate vast amounts of genomic data and has revolutionized the study
of genomes (Metzker, 2009). Large datasets combined with increased error
rates---compared to Sanger sequencing---make bioinformatic pipelines an
important aspect of research using HTS. Many bioinformatic tools exist,
and new programs that are more accurate and computationally efficient
are constantly being developed. To test these tools against known
parameter values, in silico simulation of genomic data is needed.

Although there are many sequence simulators currently available
(reviewed in Escalona, Rocha, \& Posada, 2016), most have only
rudimentary ways to generate population-level data. Events like
population-size changes, selection, or population structure can
drastically change null expectations for sequence data, but including
these possibilities is impossible with most current methods.

In the present paper I introduce \texttt{gemino}, the first available
HTS simulator in the R (R Core Team, 2018) environment. Designed for
efficient memory use, flexibility, and speed, \texttt{gemino} combines
the functionality of an HTS simulator with that of a molecular
phylogenetics simulator. Genomes can be derived from FASTA files or
simulated in silico. \texttt{gemino} can create variants from the
reference genome based on basic population-genomic summary statistics,
phylogenies, Variant Call Format (VCF) files, or coalescent simulations.
These variants are simulated based on several popular
molecular-evolution models. Another difference from other HTS simulators
is that, in addition to simulating sequencing error and mapping
qualities, \texttt{gemino} explicitly simulates variance in coverage
among sites. This makes it possible to test bioinformatic pipelines that
calculate allele frequencies. \texttt{gemino} can simulate single or
paired-ended reads for WGS on the Illumina platform, and can be extended
to simulate restriction-enzyme-associated sequencing methods and
different sequencing platforms, including those from Pacific Biosciences
(PacBio) and Oxford Nanopore Technologies (Nanopore).

After outlining the methods, I demonstrate the usefulness of
\texttt{gemino} in informing study design through three common usage
examples.

\hypertarget{features-and-methods}{%
\subsection{Features and methods}\label{features-and-methods}}

Most code is written in C++ and interfaces with R using the
\texttt{Rcpp} package (Eddelbuettel \& François, 2011). I used OpenMP to
allow for parallel processing and the \texttt{PCG} family of
pseudo-random number generators (O'Neill, 2014). An overview of the
methods are show in Figure @ref(fig:gemino-overview).

\hypertarget{read-and-create-genomes}{%
\subsubsection{Read and create genomes}\label{read-and-create-genomes}}

Haploid reference genomes can be input from FASTA files using the
function \texttt{read\_fasta}, which also accepts a FASTA index
file---created using \texttt{samtools\ faidx}---for faster processing.
Both FASTA and index files can be either uncompressed or compressed
using \texttt{gzip}. If a reference genome is not available, the
\texttt{create\_genome} function creates a reference genome of given
equilibrium nucleotide distributions, and mean and standard deviation of
the sequence-length distribution. I draw sequence lengths from a gamma
distribution (X. Li et al., 2011).

\hypertarget{alter-genomes}{%
\subsubsection{Alter genomes}\label{alter-genomes}}

Using function \texttt{filter\_sequences}, genomes can be filtered by
minimum sequence size or by the smallest sequence that retains a given
proportion of total reference sequence if sequences are sorted by
descending size. Function \texttt{merge\_sequences} shuffles reference
sequences and merges them into one.

\hypertarget{add-variants}{%
\subsubsection{Add variants}\label{add-variants}}

Variants from the reference genome are generated as haploid genomes. For
diploid organisms, two haploid variants can be randomly assigned as
being from one individual, or haploid variants can be assigned to
individuals by the user. Variants can be created in three different
ways, all of which are encompassed in the \texttt{create\_variants}
function.

(1) Mutations are randomly distributed throughout the genome based on
population-genomic statistics. Watterson's estimator (1975) informs the
number of segregating sites and Nei and Li's \(\theta\) (1979) informs
the nucleotide diversity at segregating sites. I used ``Algorithm D'' by
Vitter (1984) for sampling locations on genomes without replacement.

(2) Mutations are derived directly from a variant call format (VCF)
file. Reading and writing VCF files uses package \texttt{vcfR} (Knaus \&
Grünwald, 2017).

(3) Variant sequences are simulated along a phylogenetic tree or based
on coalescent simulation output. I used the TN93 model of nucleotide
substitution (Tamura \& Nei, 1993). Models JC69 (Jukes \& Cantor, 1969),
K80 (Kimura, 1980), F81 (Felsenstein, 1981), HKY85 (Hasegawa, Kishino,
\& Yano, 1985; Hasegawa, Yano, \& Kishino, 1984), and most others in
frequent use are special cases of the TN93 model (Yang, 2006). I
incorporated indels (insertions and deletions) of any size into the TN93
model, resulting in the following substitution rate matrix
\(\mathbf{Q}\):

\[
\mathbf{Q} = 
\begin{bmatrix}
-(\alpha_1\pi_C + \beta \pi_R + \xi) & \alpha_1 \pi_C                 & \beta \pi_A                    & \beta \pi_G \\
\alpha_1 \pi_T                 & -(\alpha_1\pi_T + \beta \pi_R + \xi) & \beta \pi_A                    & \beta \pi_G \\
\beta \pi_T                    & \beta \pi_C                    & -(\alpha_2\pi_G + \beta \pi_Y + \xi) & \alpha_2 \pi_G \\
\beta \pi_T                    & \beta \pi_C                    & \alpha_2 \pi_A                 & -(\alpha_2\pi_A + \beta \pi_Y + \xi)
\end{bmatrix}
\]

where rows specify rates of change from T, C, A, and G, respectively;
\(\pi_T\), \(\pi_C\), \(\pi_A\), and \(\pi_G\) represent substitution
rates and equilibrium frequencies; \(\pi_Y\) is the frequency of
pyrimidines (i.e., \(\pi_T + \pi_C\)); \(\pi_R\) is the frequency of
purines (i.e., \(\pi_A + \pi_G\)); \(\alpha_1\) and \(\alpha_2\) are the
rates of C\(\leftrightarrow\)T and A\(\leftrightarrow\)G transitions,
respectively; \(\beta\) is the rate of transversions; and \(\xi\) is the
summed rates of indels of all sizes.

The substitution-transition matrix (\(\mathbf{M}\)), giving the
probabilities of each substitution given that a mutation occurs, is

\[
\mathbf{M} = 
\begin{bmatrix}
0                  & \frac{q_{TC}}{q_T} & \frac{q_{TA}}{q_T} & \frac{q_{TG}}{q_T} \\
\frac{q_{CT}}{q_C} & 0                  & \frac{q_{CA}}{q_C} & \frac{q_{CG}}{q_C} \\
\frac{q_{AT}}{q_A} & \frac{q_{AC}}{q_A} & 0                  & \frac{q_{AG}}{q_A} \\
\frac{q_{GT}}{q_G} & \frac{q_{GC}}{q_G} & \frac{q_{GA}}{q_G} & 0
\end{bmatrix}
\]

where \(q_{ij} = \mathbf{Q}_{ij}\) and \(q_i = -q_{ii}\) (Yang, 2006).
The sum of values in each row \(i\) in \(\mathbf{M}\) equals
\(1 - \xi / q_i\).

The vector of indel rates by size (\(\mathbf{\Xi}\)) is

\[
\mathbf{\Xi} =
\begin{bmatrix}
    \xi_{(I)1} & \ldots  & \xi_{(I)n_I} & \xi_{(D)1} & \ldots  & \xi_{(D)n_D}
\end{bmatrix}
\]

where the first subscript indicates whether the rate pertains to an
insertion (\(I\)) or deletion (\(D\)), the second subscript indicates
size, and \(n_I\) and \(n_D\) represent the maximum sizes for insertions
and deletions respectively. All indel rates are the same among
nucleotides, and \(\sum\mathbf{\Xi} = \xi\). Rates in \(\mathbf{\Xi}\)
can be independently set to any value. Default values for insertions and
deletions are the same and are calculated, for size \(l\) from 1 to 10
(Albers et al., 2010), as such:

\[
\xi_l = R \left( \frac{\alpha_1 + \alpha_2}{2} + \beta \right)
    \frac{ \text{e}^{-l} }{ \sum^{10}_{k=1}{ \text{e}^{-k} } }
\]

where \(R\) is the average rate of indels to substitutions in eukaryotes
from Sung et al. (2016).

When choosing a substition or indel for a mutation at a position
containing nucleotide \(i\), I weight sampling based on a vector
\(\mathbf{W}_i\) consisting of row \(i\) in \(\mathbf{M}\) truncated
with \(\mathbf{\Xi} \cdot {q_i}^{-1}\). For example, if the mutation
occurs at a position containing T, the sampling weight vector would be

\[
\mathbf{W}_T = 
\begin{bmatrix}
    0 & \frac{q_{TC}}{q_T} & \frac{q_{TA}}{q_T} & \frac{q_{TG}}{q_T} & 
    \frac{\xi_{(I)1}}{q_T} & \ldots  & \frac{\xi_{(I)n_I}}{q_T} & \frac{\xi_{(D)1}}{q_T} & \ldots  & \frac{\xi_{(D)n_D}}{q_T}
\end{bmatrix}
\]

If the \(j^{\text{th}}\) item is selected, it is a substition to
nucleotide \(j\) if \(j \le 4\), an insertion of length \(j-4\) if
\(4 < j \le n_I + 4\), and a deletion of length \(j - n_I - 4\) if
\(n_I + 4 < j\). To improve the efficiency of weighted sampling from a
potentially high number of items within \(\mathbf{W}_i\), I used alias
sampling (Kronmal \& Peterson, 1979; Walker, 1974).

Phylogenetic trees are accepted as \texttt{phylo} objects from the
\texttt{ape} package (Paradis, Claude, \& Strimmer, 2004) or from
\texttt{ms}-style output from coalescent simulators such as
\texttt{scrm} (Staab, Zhu, Metzler, \& Lunter, 2015), \texttt{msms}
(Ewing \& Hermisson, 2010), \texttt{ms} (Hudson, 2002), or
\texttt{msprime} (Kelleher, Etheridge, \& McVean, 2016).
\texttt{ms}-style output can include multiple trees per variant if
recombination is included in the simulations; this is parsed properly in
\texttt{gemino}. Reference sequences are simulated along tree branches
by calculating exponential wait times for the Markov ``jump'' chain for
each sequence, where the rate of the exponential distribution is the sum
of mutation rates for each nucleotide in the sequence (Yang, 2006). The
mutation rate for nucleotide \(i\) is \(q_i\). At each jump, a position
on the sequence is sampled with a probability proportional to the
mutation rate for that nucleotide, using weighted reservoir sampling
with exponential jumps (Efraimidis \& Spirakis, 2006). A mutation type
is then sampled with probabilities from \(\mathbf{W}_i\). Jumps are
performed until the sum length of all jumps is greater than the branch
length.

\texttt{gemino} can use coalescent-simulation output that does not
include phylogenetic tree(s) if they include segregating sites. In this
case, each mutation is simply sampled from \(\mathbf{W}_i\), where \(i\)
is the nucleotide at the segregating-site position.

\hypertarget{digest-genomes}{%
\subsubsection{Digest genomes}\label{digest-genomes}}

An unlimited number of restriction-enzyme binding sites of varying
lengths can be used to digest either a reference genome or a set of
variants from the reference, using function \texttt{digest\_genome}.
Digesting variants explicitly simulates polymorphisms in restriction
enzyme binding sites, an important source of potential bias for RADseq
data (Andrews, Good, Miller, Luikart, \& Hohenlohe, 2016).

\hypertarget{simulate-sequencing-data}{%
\subsubsection{Simulate sequencing
data}\label{simulate-sequencing-data}}

Multiple sources cause variation in sequencing depth for WGS and RADseq
(Andrews et al., 2016; Escalona et al., 2016). \texttt{gemino} includes
explicit simulation of PCR, fragment size selection, and DNA shearing.
These steps occur on the fly as FASTQ files are output, so these methods
are all provided inside the \texttt{sequence} function.

PCR is modeled as a Galton--Watson discrete time branching process
(Kebschull \& Zador, 2015) for the number of fragments (\(N\)) after
\(j\) rounds of PCR:

\[
N_j = N_{j-1} + \text{Bin}(N_{j-1}, p)
\]

where \(p\) is the probability of the fragment being duplicated during
each round of PCR. The expected value and variance in \(N\) after \(j\)
rounds of PCR is as follows (Athreya \& Ney, 1972):

\[
\begin{aligned}
    \text{E}(N) &= N_0 (1 + p)^j \\
    \text{Var}(N) &= N_0 (1-p) (1+p)^{j-1} \left[ (1+p)^j - 1 \right]
\end{aligned}
\]

where \(N_0\) is the starting number of fragments. With even moderate
numbers of starting fragments, the distribution of fragment numbers
approaches normal (Figure @ref(fig:pcr-plot)). However, if starting with
only 1 of each unique fragment (which most closely resembles WGS and, to
a lesser extent, original RADseq), the distribution is neither normal
nor unimodal. For this reason, direct simulation of this branching
process is provided for WGS and original RADseq. Arguments are provided
to specify PCR bias due to fragment size and GC content.

Size selection, by default, simply filters fragments outside a range.
Alternatively, the user can assign sizes a probability of being
selected. Shearing is done by first sampling a sheared fragment size.
Then locations in the genome are sampled, which can be done with or
without weighting by GC content.

Sequencing error and quality scores for paired- or single-end Illumina
reads are based on methods by Holtgrewe (2010). Default values are
provided, but the user can specify error rates and quality-score
distribution parameters (for a normal distribution) for each position in
output reads. Although indels are rare in Illumina sequencing (Hu et
al., 2012), they can be a large source of error in reads from PacBio or
Nanopore (Kircher \& Kelso, 2010; Ono, Asai, \& Hamada, 2012; Robasky,
Lewis, \& Church, 2013). Indels are disabled by default in
\texttt{gemino}, but indel rates can be set and affected by fragment
size and GC content.

\hypertarget{writing-output}{%
\subsubsection{Writing output}\label{writing-output}}

Reference genome sequences can be written to FASTA files and sets of
variants to VCF files using the \texttt{write} function. Sequencing
reads are written to FASTQ files using the \texttt{sequence} function.
Each of these functions can output to uncompressed or gzipped files.

\hypertarget{performance}{%
\subsection{Performance}\label{performance}}

Performance was tested on a 2013 MacBook Pro running macOS High Sierra
with a 2.6GHz Intel Core i5 processor and 8 GB RAM. I compared the
performance of \texttt{gemino} functions to those in other packages
based on the time and maximum RAM required for each to perform its task.
RAM usage is presented as the maximum usage during an R session where
the only actions taken were to load the necessary package and run the
function. Thus overhead associated with loading a package is included.
For functions in packages outside \texttt{gemino}, I use the R
convention of displaying them as the package name, two colons, then the
function name (e.g., \texttt{Package::Function}).

\hypertarget{reference-genome-creating-reading-and-digesting}{%
\subsubsection{Reference-genome creating, reading, and
digesting}\label{reference-genome-creating-reading-and-digesting}}

For performance comparisons in creating, reading, and digesting
reference genomes, I used the R packages \texttt{SimRAD} and
\texttt{ShortRead} . \texttt{SimRAD} is designed to assist in research
design for RADseq studies, and \texttt{ShortRead} performs data input in
the Bioconductor system (Huber et al., 2015).

For the performance of creating a reference genome, I compared
\texttt{create\_genome} to \texttt{SimRAD::sim.DNAseq} in creating one
100 Mbp chromosome. Although \texttt{create\_genome} can use multiple
cores, I only use one for the comparison test because
\texttt{SimRAD::sim.DNAseq} can only use one and because multithreading
in \texttt{create\_genome} is done across multiple sequences. I also
report on the performance of \texttt{create\_genome} in creating a 1 Gbp
genome split among eight chromosomes.

To test reference-genome read and digestion performance, I used the
threespine stickleback (\emph{Gasterosteus aculeatus}) genome (438 Mbp;
Peichel, Sullivan, Liachko, \& White, 2017). For reading, I compared
\texttt{read\_fasta} (with and without a fasta index file) to
\texttt{ShortRead::readFasta}, separately for uncompressed and gzipped
files. I compared digestion between \texttt{digest\_genome} and
\texttt{SimRAD::insilico.digest} using the restriction enzyme
\emph{ApeKI} (1.3 million cut sites).

Table @ref(tab:SimRAD-comp-table) shows that \texttt{gemino} outperforms
\texttt{SimRAD} across all functions for both elapsed time and RAM used.
\texttt{gemino} outperformed \texttt{ShortRead} in reading FASTA files
when an index file was used in the former, but \texttt{gemino} became
slower without an index file. RAM usage for
\texttt{ShortRead::readFasta} was higher than for \texttt{read\_fasta}
largely due to the overhead associated with loading required packages.
Before running any functions, loading \texttt{SimRAD} or
\texttt{ShortRead} increased RAM usage by \textasciitilde{}325 MB, while
\texttt{gemino} only increased it by \textasciitilde{}25 MB.

\hypertarget{molecular-evolution-simulation}{%
\subsubsection{Molecular evolution
simulation}\label{molecular-evolution-simulation}}

Here I will compare my method of simulating variants with those from
package \texttt{phylosim}. However, I am not finished coding this
portion of the package.

\hypertarget{example-usage}{%
\subsection{Example usage}\label{example-usage}}

\hypertarget{choose-a-restriction-enzyme-for-radseq}{%
\subsubsection{Choose a restriction enzyme for
RADseq}\label{choose-a-restriction-enzyme-for-radseq}}

Many common restriction enzymes are already programmed into the
\texttt{digest} command, so comparing different digestions is simple.
The example below digests the stickleback genome using \emph{ApeKI} and
\emph{AscI} and plots the resulting fragment sizes (Figure
@ref(fig:choose-enzyme-run)).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{genome <-}\StringTok{ }\KeywordTok{read_fasta}\NormalTok{(}\StringTok{'stickleback_genome.fa'}\NormalTok{)}
\NormalTok{dig_ApeKI <-}\StringTok{ }\KeywordTok{digest}\NormalTok{(genome, }\StringTok{'ApeKI'}\NormalTok{, }\DataTypeTok{n_cores =} \DecValTok{4}\NormalTok{)}
\NormalTok{dig_AscI <-}\StringTok{ }\KeywordTok{digest}\NormalTok{(genome, }\StringTok{'AscI'}\NormalTok{, }\DataTypeTok{n_cores =} \DecValTok{4}\NormalTok{)}
\KeywordTok{plot_digest}\NormalTok{(}\KeywordTok{list}\NormalTok{(dig_ApeKI, dig_AscI), genome)}
\end{Highlighting}
\end{Shaded}

\hypertarget{test-variant-calling-effectiveness-in-wgs-versus-radseq}{%
\subsubsection{Test variant-calling effectiveness in WGS versus
RADseq}\label{test-variant-calling-effectiveness-in-wgs-versus-radseq}}

This portion is not yet coded.

\hypertarget{compute-the-power-of-various-depths-of-coverage-in-wgs}{%
\subsubsection{Compute the power of various depths of coverage in
WGS}\label{compute-the-power-of-various-depths-of-coverage-in-wgs}}

This portion is not yet coded.

\hypertarget{conclusion}{%
\subsection{Conclusion}\label{conclusion}}

\texttt{gemino} outperforms current programs while providing a more
flexible platform. This package should inform research design for
projects employing HTS, particularly those in population genomics.
Output from \texttt{gemino} will help develop more specific sequencing
goals in funding applications and estimate the power of a given
sequencing design. Furthermore, \texttt{gemino} can be used to test
bioinformatic pipelines under assumptions of much more complex
demographic histories than current HTS simulation platforms allow.

\hypertarget{references}{%
\section*{References}\label{references}}
\addcontentsline{toc}{section}{References}

\hypertarget{refs}{}
\leavevmode\hypertarget{ref-Albers_2010}{}%
Albers, C. A., Lunter, G., MacArthur, D. G., McVean, G., Ouwehand, W.
H., \& Durbin, R. (2010). Dindel: Accurate indel calls from short-read
data. \emph{Genome Research}, \emph{21}(6), 961--973.
\url{https://doi.org/10.1101/gr.112326.110}

\leavevmode\hypertarget{ref-Andrews_2016}{}%
Andrews, K. R., Good, J. M., Miller, M. R., Luikart, G., \& Hohenlohe,
P. A. (2016). Harnessing the power of RADseq for ecological and
evolutionary genomics. \emph{Nature Reviews Genetics}, \emph{17}(2),
81--92. \url{https://doi.org/10.1038/nrg.2015.28}

\leavevmode\hypertarget{ref-Athreya_1972}{}%
Athreya, K. B., \& Ney, P. E. (1972). The Galton-Watson process. In J.
L. Doob (Ed.), \emph{Branching processes} (pp. 1--65). Berlin: Springer.

\leavevmode\hypertarget{ref-Eddelbuettel_2011}{}%
Eddelbuettel, D., \& François, R. (2011). Rcpp: Seamless R and C++
integration. \emph{Journal of Statistical Software}, \emph{40}(8),
1--18. \url{https://doi.org/10.18637/jss.v040.i08}

\leavevmode\hypertarget{ref-Efraimidis_2006}{}%
Efraimidis, P. S., \& Spirakis, P. G. (2006). Weighted random sampling
with a reservoir. \emph{Information Processing Letters}, \emph{97}(5),
181--185. \url{https://doi.org/10.1016/j.ipl.2005.11.003}

\leavevmode\hypertarget{ref-Escalona_2016}{}%
Escalona, M., Rocha, S., \& Posada, D. (2016). A comparison of tools for
the simulation of genomic next-generation sequencing data. \emph{Nature
Reviews Genetics}, \emph{17}(8), 459--469.
\url{https://doi.org/10.1038/nrg.2016.57}

\leavevmode\hypertarget{ref-Ewing_2010}{}%
Ewing, G., \& Hermisson, J. (2010). MSMS: A coalescent simulation
program including recombination, demographic structure and selection at
a single locus. \emph{Bioinformatics}, \emph{26}(16), 2064--2065.
\url{https://doi.org/10.1093/bioinformatics/btq322}

\leavevmode\hypertarget{ref-Felsenstein_1981}{}%
Felsenstein, J. (1981). Evolutionary trees from DNA sequences: A maximum
likelihood approach. \emph{Journal of Molecular Evolution},
\emph{17}(6), 368--376. \url{https://doi.org/10.1007/bf01734359}

\leavevmode\hypertarget{ref-Hasegawa_1985}{}%
Hasegawa, M., Kishino, H., \& Yano, T.-a. (1985). Dating of the
human-ape splitting by a molecular clock of mitochondrial DNA.
\emph{Journal of Molecular Evolution}, \emph{22}(2), 160--174.
\url{https://doi.org/10.1007/bf02101694}

\leavevmode\hypertarget{ref-Hasegawa_1984}{}%
Hasegawa, M., Yano, T.-a., \& Kishino, H. (1984). A new molecular clock
of mitochondrial DNA and the evolution of hominoids. \emph{Proceedings
of the Japan Academy, Series B}, \emph{60}(4), 95--98.
\url{https://doi.org/10.2183/pjab.60.95}

\leavevmode\hypertarget{ref-Mason}{}%
Holtgrewe, M. (2010). \emph{Mason--a read simulator for second
generation sequencing data}. Institut für Mathematik und Informatik,
Freie Universität Berlin.

\leavevmode\hypertarget{ref-Hu_2012}{}%
Hu, X., Yuan, J., Shi, Y., Lu, J., Liu, B., Li, Z., \ldots{} Fan, W.
(2012). pIRS: Profile-based illumina pair-end reads simulator.
\emph{Bioinformatics}, \emph{28}(11), 1533--1535.
\url{https://doi.org/10.1093/bioinformatics/bts187}

\leavevmode\hypertarget{ref-Huber_2015}{}%
Huber, W., Carey, V. J., Gentleman, R., Anders, S., Carlson, M.,
Carvalho, B. S., \ldots{} Morgan, M. (2015). Orchestrating
high-throughput genomic analysis with bioconductor. \emph{Nature
Methods}, \emph{12}(2), 115--121.
\url{https://doi.org/10.1038/nmeth.3252}

\leavevmode\hypertarget{ref-Hudson_2002}{}%
Hudson, R. R. (2002). Generating samples under a wright-fisher neutral
model of genetic variation. \emph{Bioinformatics}, \emph{18}(2),
337--338. \url{https://doi.org/10.1093/bioinformatics/18.2.337}

\leavevmode\hypertarget{ref-JC69}{}%
Jukes, T. H., \& Cantor, C. R. (1969). Evolution of protein molecules.
In H. N. Munro (Ed.), \emph{Mammalian protein metabolism} (Vol. 3, pp.
21--131). New York: Academic Press.

\leavevmode\hypertarget{ref-Kebschull_2015}{}%
Kebschull, J. M., \& Zador, A. M. (2015). Sources of PCR-induced
distortions in high-throughput sequencing data sets. \emph{Nucleic Acids
Research}, gkv717. \url{https://doi.org/10.1093/nar/gkv717}

\leavevmode\hypertarget{ref-Kelleher_2016}{}%
Kelleher, J., Etheridge, A. M., \& McVean, G. (2016). Efficient
coalescent simulation and genealogical analysis for large sample sizes.
\emph{PLOS Computational Biology}, \emph{12}(5), e1004842.
\url{https://doi.org/10.1371/journal.pcbi.1004842}

\leavevmode\hypertarget{ref-Kimura_1980}{}%
Kimura, M. (1980). A simple method for estimating evolutionary rates of
base substitutions through comparative studies of nucleotide sequences.
\emph{Journal of Molecular Evolution}, \emph{16}(2), 111--120.
\url{https://doi.org/10.1007/bf01731581}

\leavevmode\hypertarget{ref-Kircher_2010}{}%
Kircher, M., \& Kelso, J. (2010). High-throughput DNA sequencing -
concepts and limitations. \emph{BioEssays}, \emph{32}(6), 524--536.
\url{https://doi.org/10.1002/bies.200900181}

\leavevmode\hypertarget{ref-Knaus_2017}{}%
Knaus, B. J., \& Grünwald, N. J. (2017). VCFR: A package to manipulate
and visualize variant call format data in R. \emph{Molecular Ecology
Resources}, \emph{17}(1), 44--53. Retrieved from
\url{http://dx.doi.org/10.1111/1755-0998.12549}

\leavevmode\hypertarget{ref-Kronmal_1979}{}%
Kronmal, R. A., \& Peterson, A. V. (1979). On the alias method for
generating random variables from a discrete distribution. \emph{The
American Statistician}, \emph{33}(4), 214--218.
\url{https://doi.org/10.1080/00031305.1979.10482697}

\leavevmode\hypertarget{ref-Li_2011}{}%
Li, X., Zhu, C., Lin, Z., Wu, Y., Zhang, D., Bai, G., \ldots{} Yu, J.
(2011). Chromosome size in diploid eukaryotic species centers on the
average length with a conserved boundary. \emph{Molecular Biology and
Evolution}, \emph{28}(6), 1901--1911.
\url{https://doi.org/10.1093/molbev/msr011}

\leavevmode\hypertarget{ref-Metzker_2009}{}%
Metzker, M. L. (2009). Sequencing technologies the next generation.
\emph{Nature Reviews Genetics}, \emph{11}(1), 31--46.
\url{https://doi.org/10.1038/nrg2626}

\leavevmode\hypertarget{ref-Nei}{}%
Nei, M., \& Li, W. H. (1979). Mathematical model for studying genetic
variation in terms of restriction endonucleases. \emph{Proceedings of
the National Academy of Sciences of the USA}, \emph{76}(10), 5269--5273.

\leavevmode\hypertarget{ref-Ono_2012}{}%
Ono, Y., Asai, K., \& Hamada, M. (2012). PBSIM: PacBio reads
simulatortoward accurate genome assembly. \emph{Bioinformatics},
\emph{29}(1), 119--121.
\url{https://doi.org/10.1093/bioinformatics/bts649}

\leavevmode\hypertarget{ref-Oneill_2014pcg}{}%
O'Neill, M. E. (2014). \emph{PCG: a family of simple fast
space-efficient statistically good algorithms for random number
generation}. Claremont, CA: Harvey Mudd College.

\leavevmode\hypertarget{ref-Paradis_2004}{}%
Paradis, E., Claude, J., \& Strimmer, K. (2004). APE: analyses of
phylogenetics and evolution in R language. \emph{Bioinformatics},
\emph{20}, 289--290.

\leavevmode\hypertarget{ref-Peichel_2017}{}%
Peichel, C. L., Sullivan, S. T., Liachko, I., \& White, M. A. (2017).
Improvement of the threespine stickleback genome using a hi-c-based
proximity-guided assembly. \emph{Journal of Heredity}, \emph{108}(6),
693--700. \url{https://doi.org/10.1093/jhered/esx058}

\leavevmode\hypertarget{ref-R_Core_Team_2018}{}%
R Core Team. (2018). \emph{R: a language and environment for statistical
computing}. Vienna, Austria: R Foundation for Statistical Computing.
Retrieved from \url{https://www.R-project.org/}

\leavevmode\hypertarget{ref-Robasky_2013}{}%
Robasky, K., Lewis, N. E., \& Church, G. M. (2013). The role of
replicates for error mitigation in next-generation sequencing.
\emph{Nature Reviews Genetics}, \emph{15}(1), 56--62.
\url{https://doi.org/10.1038/nrg3655}

\leavevmode\hypertarget{ref-Staab_2015}{}%
Staab, P. R., Zhu, S., Metzler, D., \& Lunter, G. (2015). Scrm:
Efficiently simulating long sequences using the approximated coalescent
with recombination. \emph{Bioinformatics}, \emph{31}(10), 1680--1682.
\url{https://doi.org/10.1093/bioinformatics/btu861}

\leavevmode\hypertarget{ref-Sung_2016}{}%
Sung, W., Ackerman, M. S., Dillon, M. M., Platt, T. G., Fuqua, C.,
Cooper, V. S., \& Lynch, M. (2016). Evolution of the insertion-deletion
mutation rate across the tree of life. \emph{G3:
Genes\(\vert\)Genomes\(\vert\)Genetics}, \emph{6}(8), 2583--2591.
\url{https://doi.org/10.1534/g3.116.030890}

\leavevmode\hypertarget{ref-TN93}{}%
Tamura, K., \& Nei, M. (1993). Estimation of the number of nucleotide
substitutions in the control region of mitochondrial dna in humans and
chimpanzees. \emph{Molecular Biology and Evolution}, \emph{10}(3),
512--526.

\leavevmode\hypertarget{ref-Vitter_1984}{}%
Vitter, J. S. (1984). Faster methods for random sampling.
\emph{Communications of the ACM}, \emph{27}(7), 703--718.
\url{https://doi.org/10.1145/358105.893}

\leavevmode\hypertarget{ref-Walker_1974}{}%
Walker, A. (1974). New fast method for generating discrete random
numbers with arbitrary frequency distributions. \emph{Electronics
Letters}, \emph{10}(8), 127. \url{https://doi.org/10.1049/el:19740097}

\leavevmode\hypertarget{ref-Watterson}{}%
Watterson, G. A. (1975). On the number of segregating sites in genetical
models without recombination. \emph{Theoretical Population Biology},
\emph{7}(2), 256--276.

\leavevmode\hypertarget{ref-Yang_2006}{}%
Yang, Z. (2006). \emph{Computational molecular evolution}. New York, NY,
USA: Oxford University Press.
\url{https://doi.org/10.1093/acprof:oso/9780198567028.001.0001}


\end{document}
