\PassOptionsToPackage{unicode=true}{hyperref} % options for packages loaded elsewhere
%DIF LATEXDIFF DIFFERENCE FILE
%DIF DEL _manuscript_orig.tex   Fri Nov 22 11:58:36 2019
%DIF ADD _manuscript.tex        Fri Nov 22 12:46:50 2019
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames*,x11names*}{xcolor}
%
\documentclass[12pt,]{article}
\usepackage{lmodern}
\usepackage{setspace}
\setstretch{2}
\usepackage{bbm}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provides euro and other symbols
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{unicode-math}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage[]{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\usepackage{xcolor}
\usepackage{hyperref}
\hypersetup{
            pdftitle={jackalope: a swift, versatile phylogenomic and high-throughput sequencing simulator},
            colorlinks=true,
            linkcolor=Blue,
            citecolor=Blue,
            urlcolor=Blue,
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage[left=2.5cm,right=2.5cm,top=3cm,bottom=3cm]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\newenvironment{Shaded}{}{}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.00,0.50,0.00}{#1}}
\newcommand{\AttributeTok}[1]{#1}
\newcommand{\BaseNTok}[1]{#1}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.00,0.50,0.50}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.00,0.50,0.00}{#1}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.00,0.50,0.00}{#1}}
\newcommand{\ConstantTok}[1]{#1}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.00,1.00}{#1}}
\newcommand{\DataTypeTok}[1]{#1}
\newcommand{\DecValTok}[1]{#1}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.00,0.50,0.00}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{#1}
\newcommand{\FunctionTok}[1]{#1}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.00,0.50,0.00}{#1}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.00,1.00}{#1}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{#1}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{1.00,0.25,0.00}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{1.00,0.25,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.50,0.50}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.00,0.50,0.50}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.00,0.50,0.50}{#1}}
\newcommand{\VariableTok}[1]{#1}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.00,0.50,0.50}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.00,0.50,0.00}{\textbf{#1}}}
\usepackage[labelfont=bf]{caption}
    \usepackage{longtable,booktabs}
            % Fix footnotes in tables (requires footnote package)
        \IfFileExists{footnote.sty}{\usepackage{footnote}\makesavenoteenv{longtable}}{}
    \usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

% set default figure and tabke placement
\usepackage{float}
\makeatletter
\def\fps@figure{H}
\def\fps@table{H}
\makeatother

% Add S to the beginning of figure and table labels if it's supplemental section


\usepackage{authblk,etoolbox}
\renewcommand\Affilfont{\small}
\makeatletter
% % patch \maketitle so that it doesn't center
% \patchcmd{\@maketitle}{center}{flushleft}{}{}
% \patchcmd{\@maketitle}{center}{flushleft}{}{}

% patch the patch by authblk so that the author block is flush left
\def\maketitle{{%
  \renewenvironment{tabular}[2][]
    {\begin{flushleft}}
    {\end{flushleft}}
  \AB@maketitle}}
\makeatother


\providecommand{\subtitle}[1]{\Huge{#1}}

\title{
    jackalope: a swift, versatile phylogenomic and high-throughput sequencing simulator
            }
            \author[1]{Lucas A. Nell}
                \affil[1]{Department of Integrative Biology, University of Wisconsin--Madison}
    % % \author{true}
% % \date{}
\date{}

\newcommand{\mean}[1]{\text{mean}\left( #1 \right)}
\newcommand{\var}[1]{\text{var}\left( #1 \right)}

% Removing extra space around \left( and \right)
\let\originalleft\left
\let\originalright\right
\renewcommand{\left}{\mathopen{}\mathclose\bgroup\originalleft}
\renewcommand{\right}{\aftergroup\egroup\originalright}


\usepackage{multirow}
\usepackage{array}
\usepackage{makecell}
\usepackage{rotating} % To display tables in landscape
\usepackage{siunitx} % Required for alignment


\usepackage{lineno}
\linenumbers
%DIF PREAMBLE EXTENSION ADDED BY LATEXDIFF
%DIF UNDERLINE PREAMBLE %DIF PREAMBLE
\RequirePackage[normalem]{ulem} %DIF PREAMBLE
\RequirePackage{color}\definecolor{RED}{rgb}{1,0,0}\definecolor{BLUE}{rgb}{0,0,1} %DIF PREAMBLE
\providecommand{\DIFaddtex}[1]{{\protect\color{blue}\uwave{#1}}} %DIF PREAMBLE
\providecommand{\DIFdeltex}[1]{{\protect\color{red}\sout{#1}}}                      %DIF PREAMBLE
%DIF SAFE PREAMBLE %DIF PREAMBLE
\providecommand{\DIFaddbegin}{} %DIF PREAMBLE
\providecommand{\DIFaddend}{} %DIF PREAMBLE
\providecommand{\DIFdelbegin}{} %DIF PREAMBLE
\providecommand{\DIFdelend}{} %DIF PREAMBLE
%DIF FLOATSAFE PREAMBLE %DIF PREAMBLE
\providecommand{\DIFaddFL}[1]{\DIFadd{#1}} %DIF PREAMBLE
\providecommand{\DIFdelFL}[1]{\DIFdel{#1}} %DIF PREAMBLE
\providecommand{\DIFaddbeginFL}{} %DIF PREAMBLE
\providecommand{\DIFaddendFL}{} %DIF PREAMBLE
\providecommand{\DIFdelbeginFL}{} %DIF PREAMBLE
\providecommand{\DIFdelendFL}{} %DIF PREAMBLE
%DIF HYPERREF PREAMBLE %DIF PREAMBLE
\providecommand{\DIFadd}[1]{\texorpdfstring{\DIFaddtex{#1}}{#1}} %DIF PREAMBLE
\providecommand{\DIFdel}[1]{\texorpdfstring{\DIFdeltex{#1}}{}} %DIF PREAMBLE
\newcommand{\DIFscaledelfig}{0.5}
%DIF HIGHLIGHTGRAPHICS PREAMBLE %DIF PREAMBLE
\RequirePackage{settobox} %DIF PREAMBLE
\RequirePackage{letltxmacro} %DIF PREAMBLE
\newsavebox{\DIFdelgraphicsbox} %DIF PREAMBLE
\newlength{\DIFdelgraphicswidth} %DIF PREAMBLE
\newlength{\DIFdelgraphicsheight} %DIF PREAMBLE
% store original definition of \includegraphics %DIF PREAMBLE
\LetLtxMacro{\DIFOincludegraphics}{\includegraphics} %DIF PREAMBLE
\newcommand{\DIFaddincludegraphics}[2][]{{\color{blue}\fbox{\DIFOincludegraphics[#1]{#2}}}} %DIF PREAMBLE
\newcommand{\DIFdelincludegraphics}[2][]{% %DIF PREAMBLE
\sbox{\DIFdelgraphicsbox}{\DIFOincludegraphics[#1]{#2}}% %DIF PREAMBLE
\settoboxwidth{\DIFdelgraphicswidth}{\DIFdelgraphicsbox} %DIF PREAMBLE
\settoboxtotalheight{\DIFdelgraphicsheight}{\DIFdelgraphicsbox} %DIF PREAMBLE
\scalebox{\DIFscaledelfig}{% %DIF PREAMBLE
\parbox[b]{\DIFdelgraphicswidth}{\usebox{\DIFdelgraphicsbox}\\[-\baselineskip] \rule{\DIFdelgraphicswidth}{0em}}\llap{\resizebox{\DIFdelgraphicswidth}{\DIFdelgraphicsheight}{% %DIF PREAMBLE
\setlength{\unitlength}{\DIFdelgraphicswidth}% %DIF PREAMBLE
\begin{picture}(1,1)% %DIF PREAMBLE
\thicklines\linethickness{2pt} %DIF PREAMBLE
{\color[rgb]{1,0,0}\put(0,0){\framebox(1,1){}}}% %DIF PREAMBLE
{\color[rgb]{1,0,0}\put(0,0){\line( 1,1){1}}}% %DIF PREAMBLE
{\color[rgb]{1,0,0}\put(0,1){\line(1,-1){1}}}% %DIF PREAMBLE
\end{picture}% %DIF PREAMBLE
}\hspace*{3pt}}} %DIF PREAMBLE
} %DIF PREAMBLE
\LetLtxMacro{\DIFOaddbegin}{\DIFaddbegin} %DIF PREAMBLE
\LetLtxMacro{\DIFOaddend}{\DIFaddend} %DIF PREAMBLE
\LetLtxMacro{\DIFOdelbegin}{\DIFdelbegin} %DIF PREAMBLE
\LetLtxMacro{\DIFOdelend}{\DIFdelend} %DIF PREAMBLE
\DeclareRobustCommand{\DIFaddbegin}{\DIFOaddbegin \let\includegraphics\DIFaddincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFaddend}{\DIFOaddend \let\includegraphics\DIFOincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFdelbegin}{\DIFOdelbegin \let\includegraphics\DIFdelincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFdelend}{\DIFOaddend \let\includegraphics\DIFOincludegraphics} %DIF PREAMBLE
\LetLtxMacro{\DIFOaddbeginFL}{\DIFaddbeginFL} %DIF PREAMBLE
\LetLtxMacro{\DIFOaddendFL}{\DIFaddendFL} %DIF PREAMBLE
\LetLtxMacro{\DIFOdelbeginFL}{\DIFdelbeginFL} %DIF PREAMBLE
\LetLtxMacro{\DIFOdelendFL}{\DIFdelendFL} %DIF PREAMBLE
\DeclareRobustCommand{\DIFaddbeginFL}{\DIFOaddbeginFL \let\includegraphics\DIFaddincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFaddendFL}{\DIFOaddendFL \let\includegraphics\DIFOincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFdelbeginFL}{\DIFOdelbeginFL \let\includegraphics\DIFdelincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFdelendFL}{\DIFOaddendFL \let\includegraphics\DIFOincludegraphics} %DIF PREAMBLE
%DIF END PREAMBLE EXTENSION ADDED BY LATEXDIFF

\begin{document}

            \maketitle




\raggedright

\hypertarget{abstract}{%
\subsection{Abstract}\label{abstract}}

High-throughput sequencing (HTS) is central to the study of population genomics
and has an increasingly important role in constructing phylogenies.
Choices in research design for sequencing projects can include
a wide range of factors, such as sequencing platform, depth of coverage, and
bioinformatic tools.
Simulating HTS data better informs these decisions.
However, current standalone HTS simulators cannot generate genomic variants under
even somewhat complex evolutionary scenarios, which greatly reduces their usefulness
for fields such as population genomics and phylogenomics.
Here I present the R package \texttt{jackalope} that simply and efficiently simulates
(i) variants from reference genomes and
(ii) reads from both Illumina and Pacific Biosciences (PacBio) platforms.
Genomic variants can be simulated using phylogenies, gene trees,
coalescent-simulation output, population-genomic summary statistics,
and Variant Call Format (VCF) files.
\texttt{jackalope} can simulate single, paired-end, or mate-pair Illumina reads,
as well as reads from Pacific Biosciences.
These simulations include sequencing errors, mapping qualities, multiplexing,
and optical/PCR duplicates.
It can read reference genomes from FASTA files and can simulate new ones,
and all outputs can be written to standard file formats.
\texttt{jackalope} is available for Mac, Windows, and Linux systems.

\textbf{Keywords:} sequencing simulator, population genomics, phylogenomics,
Illumina, Pacific Biosciences, Pool-seq

\hypertarget{introduction}{%
\subsection{Introduction}\label{introduction}}

High-throughput sequencing (HTS) is a cost-effective approach to generate vast amounts
of genomic data and has revolutionized the study of genomes
(Metzker, 2009).
The combination of massive datasets, sequencing errors, and potentially complex
evolutionary histories make bioinformatic pipelines an important aspect of
research using HTS.
Many bioinformatic tools exist, and new programs that are more accurate and
computationally efficient are constantly being developed.
Simulation of HTS data is needed to test these tools.
Although there are many sequence simulators currently available
(reviewed in Escalona, Rocha, \& Posada, 2016),
most have only rudimentary ways to generate variants from a reference genome.
For example, both \texttt{Mason} (Holtgrewe, 2010) and \texttt{GemSIM} (McElroy, Luciani, \& Thomas, 2012) place mutations
randomly throughout genome sequences, the former based on per-site mutation rates
and the latter on a set number of mutations per variant haplotype.
In either case, \DIFdelbegin \DIFdel{these simulators }\DIFdelend \DIFaddbegin \DIFadd{simulations alone }\DIFaddend are not useful for researchers interested
in simulating HTS data to test more complex evolutionary questions, such as
(a) how demographic changes and selection interact to affect genome-wide diversity or
(b) how incongruence between gene trees might affect reconstructed phylogenies.

Recently, pipelines have been developed that use multiple programs
to simulate complex evolutionary histories and HTS on the resulting populations
or species.
\texttt{TreeToReads} (McTavish et al., 2017) simulates sequences along a single
phylogenetic tree using \texttt{Seq-Gen} (Rambaut \& \DIFdelbegin \DIFdel{Grass}\DIFdelend \DIFaddbegin \DIFadd{Grassly}\DIFaddend , 1997)
and generates Illumina reads using \texttt{ART} (Huang, Li, Myers, \& Marth, 2011).
The newer pipeline \texttt{NGSphy} (Escalona, Rocha, \& Posada, 2018) uses \texttt{INDELible} (Fletcher \& Yang, 2009)
for more mutation options than \texttt{Seq-Gen} and
can simulate along multiple gene trees via \texttt{SimPhy} (Mallo, Oliveira Martins, \& Posada, 2015).
These pipelines are quite powerful.
However, they both require the installation of multiple programs to use them,
and the vast array of features creates a steep learning curve when getting started.
Additionally, the lack of integration between inner programs means that the
entire process is not as computationally efficient as a single standalone program.

In the present paper I introduce \texttt{jackalope}, an R (R Core Team, 2019) package
that combines the functionality of an HTS simulator with that of a
phylogenomics simulator.
\DIFaddbegin \DIFadd{Although }\texttt{\DIFadd{jackalope}} \DIFadd{can process output from coalescent simulators,
it does not do coalescent simulations itself and assumes input files to be true.
}\DIFaddend I chose the R platform because of its simple integration with C++ code via
the \texttt{Rcpp} package (Eddelbuettel \& François, 2011).
The interface with R allows the user greater flexibility, while the underlying
C++ code improves memory use and speed.
\texttt{jackalope} can read genomes from FASTA files or simulate them \DIFdelbegin \DIFdel{in silico}\DIFdelend \DIFaddbegin \emph{\DIFadd{in silico}}\DIFaddend .
It can also create variants from the reference genome based on basic
population-genomic summary statistics, phylogenies, gene trees,
Variant Call Format (VCF) files, or matrices of segregating sites.
These variants can be simulated based on any of several popular
molecular-evolution models.
\texttt{jackalope} simulates single, paired-ended, or mate-pair reads on the Illumina platform,
as well as Pacific Biosciences (PacBio) reads.
All information generated by \texttt{jackalope} can be output to standard file formats.

After outlining the methods, I compare the performance of \texttt{jackalope} to that of
other popular programs.
Lastly, I demonstrate the usefulness of \texttt{jackalope} through \DIFdelbegin \DIFdel{three }\DIFdelend \DIFaddbegin \DIFadd{four }\DIFaddend usage examples.

\hypertarget{features-and-methods}{%
\subsection{Features and methods}\label{features-and-methods}}

Most code is written in C++ and interfaces with R using the \texttt{Rcpp} package
(Eddelbuettel \& François, 2011).
I used OpenMP to allow for parallel processing and
the \texttt{PCG} family of thread-safe, pseudo-random number generators
(O'Neill, 2014).
\DIFaddbegin \DIFadd{I use alias sampling (Kronmal \& Peterson, 1979; Walker, 1974) for all weighted sampling.
}\DIFaddend Package \texttt{RcppProgress} provides the thread-safe progress bar
(Forner, 2018).
All input and output files can have \texttt{gzip} or \texttt{bgzip} compression, using the
\texttt{zlib} and \texttt{htslib} (Li et al., 2009) libraries.
Access to these libraries uses the R packages
\texttt{zlibbioc} (Morgan, 2019) and \texttt{Rhtslib} (Hayden \& Morgan, 2019) to improve portability.
\DIFdelbegin \DIFdel{An overview of the methods is show in Figure \ref{fig:jackalope-overview-figure}.
}\DIFdelend 

\DIFdelbegin %DIFDELCMD < \hypertarget{the-ref_genome-class}{%
%DIFDELCMD < \subsubsection{\texorpdfstring{The \texttt{ref\_genome} class}{The ref\_genome class}}\label{the-ref_genome-class}}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{Haploid reference genomes
are represented by the class }\DIFdelend \DIFaddbegin \DIFadd{There are two classes in }\texttt{\DIFadd{jackalope}}\DIFadd{, one that stores information on reference genomes
(}\DIFaddend \texttt{ref\_genome}\DIFdelbegin \DIFdel{, an
}\DIFdelend \DIFaddbegin \DIFadd{), the other that stores information on sets of variants (}\texttt{\DIFadd{variants}}\DIFadd{).
Both classes are }\DIFaddend \texttt{R6} (Chang, 2019) \DIFdelbegin \DIFdel{class that acts as a wrapper around a pointer to an }\DIFdelend \DIFaddbegin \DIFadd{classes that wrap pointers to }\DIFaddend underlying C++ \DIFdelbegin \DIFdel{object that stores all the sequence information.
They can }\DIFdelend \DIFaddbegin \DIFadd{objects.
Those C++ objects store all the information.
Reference genomes are simply vectors of chromosome names and sequence strings.
Variants are stored as nested vectors of mutation information.
This can greatly reduce memory usage (unless the mutation density is high), and
it provides the ability to output VCF files based on variants.
I chose }\texttt{\DIFadd{R6}} \DIFadd{classes because they allowed me to make the pointers to the C++ objects
private, thereby preventing users from accidentally manipulating them.
Methods in }\texttt{\DIFadd{jackalope}}\DIFadd{'s classes allow the user to view and manipulate aspects of the
C++ objects.
Structuring the classes this way provides a high degree of flexibility and
minimizes the chances of copying large objects in memory.
An overview of how these classes are created and used to produce output files is shown
in Figure \ref{fig:jackalope-overview-figure}.
}

\hypertarget{reference-genomes}{%
\subsubsection{Reference genomes}\label{reference-genomes}}

\DIFadd{Haploid reference genomes can }\DIFaddend be generated from FASTA files\DIFdelbegin \DIFdel{using the function }\texttt{\DIFdel{read\_fasta}}%DIFAUXCMD
\DIFdel{.
This function also accepts FASTA }\DIFdelend \DIFaddbegin \DIFadd{, optionally with }\DIFaddend index
files---created using \texttt{samtools\ faidx}\DIFdelbegin \DIFdel{---for faster }\DIFdelend \DIFaddbegin \DIFadd{---that speed up }\DIFaddend processing.
If a reference genome is not available, \DIFdelbegin \DIFdel{the }\texttt{\DIFdel{create\_genome}} %DIFAUXCMD
\DIFdel{function creates
a reference genome of }\DIFdelend \DIFaddbegin \DIFadd{one can be
simulated based on }\DIFaddend given equilibrium nucleotide distributions.
\DIFdelbegin \DIFdel{Sequence }\DIFdelend \DIFaddbegin \DIFadd{Chromosome }\DIFaddend lengths are drawn from a \DIFdelbegin \DIFdel{gamma distribution
(Li et al., 2011)
}\DIFdelend \DIFaddbegin \DIFadd{Gamma distribution
}\DIFaddend with a mean and standard deviation provided by the user.
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{The access provided by the R class }\texttt{\DIFdel{ref\_genome}} %DIFAUXCMD
\DIFdel{is designed to both
maximize flexibility and minimize copying and the chances of
printing extremely large strings to the console.
Methods in }\texttt{\DIFdel{ref\_genome}} %DIFAUXCMD
\DIFdel{allow the user to view the
number of sequences, sequence sizes , sequence names, individual sequence strings,
and nucleotide proportions (GC or otherwise).Users can also edit sequence names, remove one or more sequences by name,
and add sequences manually.
Method }\texttt{\DIFdel{filter\_sequences}} %DIFAUXCMD
\DIFdel{filters genomes by the minimum sequence size or
by the smallest sequence that retains a given proportion of total reference
sequence if sequences are sorted by descending size.
Using method }\texttt{\DIFdel{merge\_sequences}}%DIFAUXCMD
\DIFdel{, users can shuffle reference sequences and
merge them into one.
Method }\texttt{\DIFdel{replace\_Ns}} %DIFAUXCMD
\DIFdel{replaces any }\texttt{\DIFdel{N}}%DIFAUXCMD
\DIFdel{s in the reference sequence with nucleotides
that are sampled with weights proportional to their equilibrium distributions
(provided by the user).
Reference genomes can be written to FASTA files using the }\texttt{\DIFdel{write\_fasta}} %DIFAUXCMD
\DIFdel{function.
}\DIFdelend \DIFaddbegin \DIFadd{The Gamma distribution was chosen because of its flexibility and due to evidence
that it works well for chromosome sizes in diploid eukaryotes (Li et al., 2011).
}\DIFaddend 

\hypertarget{creating-variants}{%
\subsubsection{Creating variants}\label{creating-variants}}

\DIFdelbegin \DIFdel{Haploid variants from the reference genome are generated using the
}\texttt{\DIFdel{create\_variants}} %DIFAUXCMD
\DIFdel{function.
To organize a potentially large amount of information that can be input to this function,
I added helper functions that handle parts of the input.
There are helper functions for the higher-level method information
(information from phylogenies, coalescent simulations, etc.) and
for various aspects of the molecular evolution (substitutions, indels, and among-site
variability in mutation rates).
The first subsection below outlines the higher-level information functions, and
the second subsection details the molecular evolution functions.
}\DIFdelend \DIFaddbegin \hypertarget{methods}{%
\paragraph{Methods}\label{methods}}
\DIFaddend 

\DIFdelbegin %DIFDELCMD < \hypertarget{higher-level-method-information}{%
%DIFDELCMD < \paragraph{Higher-level method information}%DIFDELCMD < \label{higher-level-method-information}%%%
}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend There are five ways to generate \DIFaddbegin \DIFadd{haploid }\DIFaddend variants from the reference genome, and
\DIFdelbegin \DIFdel{each has a function associated with it.
The names of these functions follow the form }\texttt{\DIFdel{vars\_X}} %DIFAUXCMD
\DIFdel{for method }\texttt{\DIFdel{X}}%DIFAUXCMD
\DIFdel{,
and }\DIFdelend information on all \DIFdelbegin \DIFdel{of the function }\DIFdelend \DIFaddbegin \DIFadd{the methods }\DIFaddend can be found in the \DIFdelbegin \texttt{\DIFdel{vars\_functions}} %DIFAUXCMD
\DIFdel{documentation .
The outputs of these functions are meant to be passed to the }\texttt{\DIFdel{vars\_info}} %DIFAUXCMD
\DIFdel{argument
in }\texttt{\DIFdel{create\_variants}}%DIFAUXCMD
\DIFdel{.
}%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{The first two methods directly specify numbers and locations of mutations and
therefore }\DIFdelend \DIFaddbegin \DIFadd{documentation (under }\texttt{\DIFadd{vars\_functions}}\DIFadd{).
The first two variant methods }\DIFaddend do not require \DIFdelbegin \DIFdel{any phylogenomic methods in }\texttt{\DIFdel{jackalope}}%DIFAUXCMD
\DIFdelend \DIFaddbegin \DIFadd{phylogenetic information}\DIFaddend .
First, a \DIFdelbegin \DIFdel{variant call format }\DIFdelend \DIFaddbegin \DIFadd{Variant Call Format }\DIFaddend (VCF) file can directly specify mutations for each variant\DIFdelbegin \DIFdel{(function }\texttt{\DIFdel{vars\_vcf}}%DIFAUXCMD
\DIFdel{)}\DIFdelend .
This method works using the \texttt{\DIFdelbegin \DIFdel{vcfR}\DIFdelend \DIFaddbegin \DIFadd{htslib}\DIFaddend } \DIFdelbegin \DIFdel{package
(Knaus \& Grünwald, 2017) }\DIFdelend \DIFaddbegin \DIFadd{C library }\DIFaddend and
is the only method that does not require any molecular-evolution information.
Second, \DIFdelbegin \DIFdel{matrices of }\DIFdelend segregating sites from coalescent output can \DIFdelbegin \DIFdel{provide }\DIFdelend \DIFaddbegin \DIFadd{inform }\DIFaddend the locations of mutations
\DIFdelbegin \DIFdel{(function }\texttt{\DIFdel{vars\_ssites}}%DIFAUXCMD
\DIFdel{).
Molecular evolution informationpassed to }\texttt{\DIFdel{create\_variants}} %DIFAUXCMD
\DIFdel{then informs the sampling
for the type of mutation at each site.
}\DIFdelend \DIFaddbegin \DIFadd{and which variants have them.
The user provides molecular evolution information, which is then used to sample
for mutation types at each segregating site.
Sampling is weighted based on each mutation type's rate.
}\DIFaddend The segregating-site information can take the form of
(1) a coalescent-simulator object from the \texttt{scrm}
(Staab, Zhu, Metzler, \& Lunter, 2015) or
\texttt{coala} (Staab \& Metzler, 2016) package, or
(2) a file containing output from a coalescent simulator in the format of the
\texttt{ms} (Hudson, 2002) or \texttt{msms} (Ewing \& Hermisson, 2010) program.

The \DIFdelbegin \DIFdel{last three methods simulate sequences along phylogenies}\DIFdelend \DIFaddbegin \DIFadd{remaining three variant methods use phylogenetic information}\DIFaddend .
In the \DIFdelbegin \DIFdel{first of these methods}\DIFdelend \DIFaddbegin \DIFadd{third method}\DIFaddend , phylogenetic tree(s) can be directly input from either
\texttt{phylo} object(s) or NEWICK file(s)\DIFdelbegin \DIFdel{(function }\texttt{\DIFdel{vars\_phylo}}%DIFAUXCMD
\DIFdel{).
One tree can be used for all genome sequences, or each sequence can use a separate tree }\DIFdelend .
\DIFdelbegin \DIFdel{In the second method}\DIFdelend \DIFaddbegin \DIFadd{These can be a single species tree or one gene tree per chromosome.
Fourth}\DIFaddend , users can pass an estimate for \(\theta\), the population-scaled mutation rate\DIFdelbegin \DIFdel{(function }\texttt{\DIFdel{vars\_theta}}%DIFAUXCMD
\DIFdel{)}\DIFdelend .
A random coalescent tree is first generated using the \texttt{rcoal} function
in the \texttt{ape} package (Paradis \& Schliep, 2018).
Then, its total tree length is scaled to be
\(\theta / \mu \sum_{i=1}^{n-1}{1 / i}\) for \(n\) variants and an equilibrium
mutation rate of \(\mu\).
The last method allows for \DIFdelbegin \DIFdel{simulation of }\DIFdelend recombination by simulating along
gene trees that can differ both within and among reference \DIFdelbegin \DIFdel{sequences
(function }\texttt{\DIFdel{vars\_gtrees}}%DIFAUXCMD
\DIFdel{)}\DIFdelend \DIFaddbegin \DIFadd{chromosomes}\DIFaddend .
Similarly to simulations using coalescent segregating sites, gene trees can be
from \texttt{scrm} or \texttt{coala} objects, or from \texttt{ms}-style output files.

\DIFdelbegin \DIFdel{For the phylogenomicmethods, variants are simulated along
tree branches by
generating exponential wait times
for the Markov ``jump'' chain for each sequence,
where the rate of the exponential distribution is the sum of mutation rates for
all nucleotides in the sequence (Yang, 2006).
At each jump, a position on the sequence is sampled with a probability proportional to the mutation rate for that position.
A position's mutation rate is the product of its ``gamma distance''
(determined by
among-site variability in mutation rates) and the overall mutation rate
for the nucleotide at that position
(determined by the summed indel and substitution rates).
After sampling a position, a mutation type is sampled with probabilities
proportional to the rate of each mutation type for }\DIFdelend \DIFaddbegin \DIFadd{In these last three ``phylogenomic'' methods, chromosomal sequences are evolved along
phylogenetic trees (either species or gene trees).
All chromosomes evolve independently.
If recombination is included, multiple gene trees are used
per chromosome, and chromosomal regions referred to by different gene trees
evolve independently.
Evolving independent chromosomal regions along phylogenetic trees starts by
using the reference genome sequences as those for the root of the tree.
Then, for each region and branch, the number of newly generated mutations is
proportional to }\DIFaddend the \DIFdelbegin \DIFdel{nucleotide
present at the sampled position.
Jumps are performed until the summed length of all jumps is
greater than the
}\DIFdelend branch length \DIFaddbegin \DIFadd{(see below for more details)}\DIFaddend .
\DIFaddbegin \DIFadd{These new mutations are then assigned to the daughter node on the tree.
This process is repeated down the tree, from root to tips.
In }\texttt{\DIFadd{jackalope}}\DIFadd{, mutation information is passed through the tree in such a
way that no intermediate objects are created.
}\DIFaddend 

\DIFdelbegin \DIFdel{Because sampling of mutation positions incorporates both among-site and among-nucleotide
variability in mutation rates,
users can separately specify rate variation that occurs due to
(1)where the nucleotides are positioned on the sequence molecule and
(2) the chemical structures of the nucleotides themselves.
To accomplish this, }\texttt{\DIFdel{jackalope}} %DIFAUXCMD
\DIFdel{first splits each sequence into
smaller regions, where each region contains the same gamma distance.
The nucleotide-level rates are then summed by region and multiplied by the region's
gamma distance to get the region's overall rate.
I used these total rates as weights to sample regions, and
I incorporated a binary search treeto speed this sampling.
Simple inversion sampling is used to sample within a region.
}%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{Mutation-type sampling is performed by creating 4 vectors (one for each nucleotide), each containing all possible mutation types:
substitutions to other nucleotides and insertions/deletions of all specified lengths.
I then use alias sampling
(Kronmal \& Peterson, 1979; Walker, 1974)
to sample within this vector.
}%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend \hypertarget{molecular-evolution}{%
\paragraph{Molecular evolution}\label{molecular-evolution}}

\DIFdelbegin \DIFdel{Molecular evolution information is provided to the }\texttt{\DIFdel{create\_variants}}
%DIFAUXCMD
\DIFdel{function through organizing functions that are specific to the type of information:
substitutions , indels (insertions and deletions),
and variation in mutation rates among sites.
Substitutions use the }\texttt{\DIFdel{sub\_models}} %DIFAUXCMD
\DIFdel{group of functions,
indels use the }%DIFDELCMD < \texttt{%%%
\DIFdelend \DIFaddbegin \DIFadd{Both substitutions and }\DIFaddend indels \DIFdelbegin %DIFDELCMD < } %%%
\DIFdel{function,
and among-site variation uses the }\texttt{\DIFdel{site\_var}} %DIFAUXCMD
\DIFdel{function.
A table of rates from
Sung et al. (2016)
is included in the package
as a guide for reasonable rates to use}\DIFdelend \DIFaddbegin \DIFadd{can be simulated by }\texttt{\DIFadd{jackalope}}\DIFadd{.
It is assumed that the substitution- and indel-producing processes are independent,
so they are simulated separately to improve computational efficiency (Fletcher \& Yang, 2009)}\DIFaddend .

\DIFdelbegin \DIFdel{Each substitution model uses its own function of the pattern }\texttt{\DIFdel{sub\_M}} %DIFAUXCMD
\DIFdel{for model }\texttt{\DIFdel{M}}%DIFAUXCMD
\DIFdel{, and the provided }\texttt{\DIFdel{sub\_models}} %DIFAUXCMD
\DIFdel{documentation includes information to help
users choose among them.
The following substitution }\DIFdelend \DIFaddbegin \hypertarget{substitutions}{%
\subparagraph{Substitutions}\label{substitutions}}

\DIFadd{For substitutions, the following }\DIFaddend models can be employed:
TN93 (Tamura \& Nei, 1993),
JC69 (Jukes \& Cantor, 1969),
K80 (Kimura, 1980),
F81 (Felsenstein, 1981),
HKY85 (Hasegawa, Kishino, \& Yano, 1985; Hasegawa, Yano, \& Kishino, 1984),
F84 (Thorne, Kishino, \& Felsenstein, 1992),
GTR (Tavar\a'e, 1986),
and UNREST (Yang, 1994).
If using the UNREST model, equilibrium nucleotide frequencies (\(\mathbf{\pi}\)) are
calculated by solving for \(\mathbf{\pi} \mathbf{Q} = 0\), where \(\mathbf{Q}\) is the
substitution rate matrix.
This is done by finding the left eigenvector of \(\mathbf{Q}\) that
corresponds to the eigenvalue closest to zero.
\DIFaddbegin \DIFadd{Each substitution model uses its own function, and the provided documentation
(under }\texttt{\DIFadd{sub\_models}}\DIFadd{) includes information to help users choose among them.
}\DIFaddend 

\DIFdelbegin \DIFdel{Providing insertion and deletion information is organized in the }\DIFdelend \DIFaddbegin \DIFadd{Substitution rates can also vary among sites.
A random-sites, discrete-Gamma model is used, where sites' rates are assumed to
be independent of one another and are derived from a Gamma distribution split into
\(K\) rate categories of equal probability (Yang, 2006).
The Gamma distribution is constrained to have a mean of 1.
A site's overall rate is its nucleotide's mutation rate multiplied by the
``Gamma distance'' associated with its rate category.
The user can also specify a proportion of invariant sites.
}

\DIFadd{For the phylogenetic methods, substitutions are simulated for each branch by first
calculating the transition-probability matrix (\(P(t)\)) for each Gamma category based on
the branch length.
Substitutions are then sampled for each site using the \(P(t)\) matrix that coincides with
the Gamma category at that site.
This procedure has been used in multiple programs to capture the substitution process,
including }\DIFaddend \texttt{\DIFdelbegin \DIFdel{indels}\DIFdelend \DIFaddbegin \DIFadd{Seq-Gen}\DIFaddend } \DIFdelbegin \DIFdel{function.
It first requires }\DIFdelend \DIFaddbegin \DIFadd{(Rambaut \& Grassly, 1997) and }\texttt{\DIFadd{INDELible}} \DIFadd{(Fletcher \& Yang, 2009).
}

\hypertarget{indels}{%
\subparagraph{Indels}\label{indels}}

\DIFadd{Information for insertions and deletions is provided separately, but
the same information is needed for both.
The first requirement is }\DIFaddend an overall rate parameter, which is for the sum among all
nucleotides; indel rates do not differ among nucleotides.
The \DIFdelbegin \texttt{\DIFdel{indels}} %DIFAUXCMD
\DIFdel{function also requires information about the }\DIFdelend relative rates of indels of different sizes \DIFdelbegin \DIFdel{, which }\DIFdelend can be provided in 3 different ways\DIFdelbegin \DIFdel{.
}\DIFdelend \DIFaddbegin \DIFadd{:
}\DIFaddend First, rates can be proportional to \(\exp(-u)\) for indel length \(u\) from
1 to the maximum possible length, \DIFdelbegin \DIFdel{\(M\)
}\DIFdelend \DIFaddbegin \DIFadd{\(L\) }\DIFaddend (Albers et al., 2010).
Second, rates can be generated from a Lavalette distribution,
where the rate for length \(u\) is proportional to
\DIFdelbegin \DIFdel{\(\left[{u M / (M - u + 1)}\right]^{-a}\)
}\DIFdelend \DIFaddbegin \DIFadd{\(\left[{u L / (L - u + 1)}\right]^{-a}\) }\DIFaddend (Fletcher \& Yang, 2009).
Third, relative rates can be specified directly by providing a length-\DIFdelbegin \DIFdel{\(M\)
}\DIFdelend \DIFaddbegin \DIFadd{\(L\)
}\DIFaddend numeric vector of positive values.
Indels up to 1 Mb are allowed.

\DIFdelbegin \DIFdel{Among-site variation in mutation rates
, specified in the }\DIFdelend \DIFaddbegin \DIFadd{Indels are generated along phylogenetic-tree branches using
the \(\tau\)-leaping approximation (Cao, Gillespie, \& Petzold, 2006; Wieder, Fink, \& Wegner, 2011) to the
Doob--Gillespie algorithm (Doob, 1942; Gillespie, 1976).
I chose the Doob--Gillespie algorithm because it only requires instantaneous rates
rather than the calculation of a transition-probability matrix, making it easily used
for indels (Yang, 2006).
This method has been used in }\DIFaddend \texttt{\DIFdelbegin \DIFdel{site\_var}\DIFdelend \DIFaddbegin \DIFadd{INDELible}\DIFaddend } \DIFdelbegin \DIFdel{function, is included either by generating gamma distances from a distribution
or by passing them manually.
Gamma distances }\DIFdelend \DIFaddbegin \DIFadd{(Fletcher \& Yang, 2009) and }\texttt{\DIFadd{Dawg}} \DIFadd{(Cartwright, 2005)
to simulate indels.
The Doob--Gillespie algorithm works by generating waiting times for each branch
length that represent the time until the next mutation occurs somewhere on the sequence.
Waiting times }\DIFaddend are generated from \DIFdelbegin \DIFdel{a Gamma }\DIFdelend \DIFaddbegin \DIFadd{an exponential }\DIFaddend distribution with a \DIFdelbegin \DIFdel{fixed mean of 1
and with a shape parameter provided by the
user;
a proportion of these regions can optionally be invariant.
Users can also pass a list of matrices, one for each reference sequence,
with a gamma distance and end point for each sequence region.
The gamma distances can optionally be written to a BED file.
}\DIFdelend \DIFaddbegin \DIFadd{rate equal to the
sum of indel rates for all nucleotides in the sequence.
Waiting times (and resulting mutations) are generated until the sum of all times
is greater than the branch length.
}\DIFaddend 

\DIFdelbegin %DIFDELCMD < \hypertarget{the-variants-class}{%
%DIFDELCMD < \subsubsection{\texorpdfstring{The \texttt{variants} class}{The variants class}}%DIFDELCMD < \label{the-variants-class}%%%
}
%DIFDELCMD < %%%
\DIFdelend \DIFaddbegin \DIFadd{The \(\tau\)-leaping approximation improves efficiency by breaking the branch length
into \(\tau\)-sized chunks and using a Poisson distribution to generate how many
mutations of each type occur in each chunk.
The maximum value of \(\tau\) is
}\DIFaddend 

\DIFdelbegin \DIFdel{Haploid variant information is stored in the }\texttt{\DIFdel{variants}} %DIFAUXCMD
\DIFdel{class.
Similarly to }\texttt{\DIFdel{ref\_genome}}%DIFAUXCMD
\DIFdel{, this R6 class wraps a pointer to a C++ object
that stores all the information, and it was designed to prevent copying of large
objects in memory.
The underlying C++ class does not store whole variant genomes, but
rather just their mutation information---this dramatically reduces memory usage.
Methods in }\texttt{\DIFdel{variants}} %DIFAUXCMD
\DIFdel{allow the user to view the number of sequences/variants,
variant sequence sizes,
sequence/variant names, and
individual variant-sequence strings.Users can also edit variant names, remove one or more variants by name, add blank variants, duplicate existing variants,
and manually add mutations.
Variant information can be written to VCF files using the }\texttt{\DIFdel{write\_vcf}} %DIFAUXCMD
\DIFdel{function, where each variant can optionally be considered one of multiple haplotypes for samples with ploidy levels \textgreater{} 1.
Alternatively, variant genomes can be written to separate FASTA files
using }\texttt{\DIFdel{write\_fasta}}%DIFAUXCMD
\DIFdelend \DIFaddbegin \begin{align*}
\DIFadd{\tau} &= \DIFadd{\min \left[ \frac { \max \left( \epsilon \, C, 1 \right) } { | \mu | }, \frac { \max \left( \epsilon \, C, 1 \right)^2 } { \sigma^2 } \right]} \\
\DIFadd{\mu }&\DIFadd{= C \; \sum_i{ m_i v_i } }\\
\DIFadd{\sigma^2 }&\DIFadd{= C \; \sum_i{ m_i v_i^2 } ~ \rlap{,}}
\end{align*}


\DIFadd{for an error control parameter \(\epsilon\) (where \(0 < \epsilon \ll 1\)),
chromosome length \(C\),
mutation rate for one indel type \(m_i\), and
effect on chromosome length for one indel type \(v_i\) (Cao et al., 2006; Wieder et al., 2011).
Increasing the error control parameter results in faster simulations that
approximate the exact Doob--Gillespie algorithm less precisely.
It can be changed by the user, but has a default value of 0.03
as recommended by Cao et al. (2006).
The rate parameter for the Poisson distribution that generates the number of
mutations over the entire chromosome across \(\tau\) time units is \(\tau C m_i\)
for indel type \(i\)}\DIFaddend .

\DIFdelbegin %DIFDELCMD < \hypertarget{simulate-sequencing-data}{%
%DIFDELCMD < \subsubsection{Simulate sequencing data}%DIFDELCMD < \label{simulate-sequencing-data}%%%
}
%DIFDELCMD < %%%
\DIFdelend \DIFaddbegin \hypertarget{simulating-sequencing-data}{%
\subsubsection{Simulating sequencing data}\label{simulating-sequencing-data}}
\DIFaddend 

Both \DIFdelbegin \DIFdel{R6 classes }\texttt{\DIFdel{ref\_genome}} %DIFAUXCMD
\DIFdel{and }\texttt{\DIFdel{variants}} %DIFAUXCMD
\DIFdel{can be input to the sequencing functions
}\texttt{\DIFdel{illumina}} %DIFAUXCMD
\DIFdel{and }\texttt{\DIFdel{pacbio}}%DIFAUXCMD
\DIFdel{.
Sequences }\DIFdelend \DIFaddbegin \DIFadd{reference genomes and variants can be simulated for Illumina and PacBio sequencing.
Chromosomes }\DIFaddend from which reads are derived are sampled with weights proportional to
their length.
If \DIFdelbegin \DIFdel{a }\texttt{\DIFdel{variants}} %DIFAUXCMD
\DIFdel{object is }\DIFdelend \DIFaddbegin \DIFadd{variants are }\DIFaddend provided, individual variants are sampled with equal probabilities
by default.
Alternatively, the user can specify sampling weights for each variant
to simulate the library containing differing amounts of DNA from each.
Both methods also allow for a probability of read duplication, which might occur
due to PCR in either method and from optical duplicates in Illumina sequencing.
\texttt{jackalope} can create reads using multiple threads by having a read ``pool'' for
each thread and having pools write to file only when they are full.
This reduces conflicts that occur when multiple threads attempt to write to disk
at the same time.
The size of a ``full'' pool can be adjusted, and larger sizes should increase both
speed and memory usage.
\DIFdelbegin \DIFdel{Reads are output to FASTQ files.
}%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{Function }\texttt{\DIFdel{illumina}} %DIFAUXCMD
\DIFdel{simulates }\DIFdelend \DIFaddbegin \DIFadd{Illumina reads can be }\DIFaddend single, paired-ended, or mate-pair\DIFdelbegin \DIFdel{Illumina reads, while }\texttt{\DIFdel{pacbio}} %DIFAUXCMD
\DIFdel{simulates reads from the Pacific Biosciences (PacBio) platform.
Illumina read simulation is }\DIFdelend \DIFaddbegin \DIFadd{, and I used
methods }\DIFaddend based on the \texttt{ART} program (Huang et al., 2011)\DIFdelbegin \DIFdel{, and
}\DIFdelend \DIFaddbegin \DIFadd{.
}\DIFaddend PacBio read simulation is based on \texttt{SimLoRD} (Stöcker, Köster, \& Rahmann, 2016).
Each was re-coded in C++ to more seemlessly integrate into \texttt{jackalope}.
Function inputs emulate the program they were based on,
and users are encouraged in each function's documentation to also cite
\texttt{ART} or \texttt{SimLoRD}.

\DIFaddbegin \hypertarget{writing-output}{%
\subsubsection{Writing output}\label{writing-output}}

\DIFadd{Reference genomes and variants can be written to FASTA files; the latter
has a separate file for each variant.
Variant information can also be written to VCF files, where each variant can
optionally be considered one of multiple haplotypes for samples with ploidy levels \textgreater{} 1.
Simulated gene trees can be written to }\texttt{\DIFadd{ms}}\DIFadd{-style output files, and phylogenies can be
written to NEWICK files using }\texttt{\DIFadd{ape::write.tree}}\DIFadd{.
All sequencing reads are output to FASTQ files.
}

\DIFaddend \hypertarget{validation-and-performance}{%
\subsection{Validation and performance}\label{validation-and-performance}}

I validated \texttt{jackalope} by conducting a series of simulations for variant creation
and Illumina and PacBio sequencing.
Using known inputs, I compared predicted to observed values of outputs,
and \texttt{jackalope} conformed closely to expectations
(Supporting Information Figures \DIFdelbegin \DIFdel{S1--S12}\DIFdelend \DIFaddbegin \DIFadd{S1--S13}\DIFaddend ).

I compared the performance of \texttt{jackalope} to existing programs on a MacBook
Pro running macOS \DIFdelbegin \DIFdel{Mojave (version 10.14.4}\DIFdelend \DIFaddbegin \DIFadd{Catalina (version 10.15}\DIFaddend ) with a 2.6GHz Intel Core i5 processor
and 8 GB RAM.
Elapsed time and maximum memory used were measured by using the GNU \texttt{time} program
(\texttt{/usr/bin/time} from the terminal).
For each test, \DIFdelbegin \DIFdel{10 }\DIFdelend \DIFaddbegin \DIFadd{5 }\DIFaddend runs of each program were tested in random order.
\DIFaddbegin \DIFadd{In the text below, numbers presented are means from these 5 tests unless otherwise stated.
}\DIFaddend 

\hypertarget{creating-variants-1}{%
\subsubsection{Creating variants}\label{creating-variants-1}}

For creating variants, I compared \texttt{jackalope} to \texttt{\DIFdelbegin \DIFdel{Seq-Gen}\DIFdelend \DIFaddbegin \DIFadd{INDELible}\DIFaddend }
(\DIFdelbegin \DIFdel{Rambaut \& Grass, 1997}\DIFdelend \DIFaddbegin \DIFadd{version 1.03; Fletcher \& Yang, 2009}\DIFaddend ).
I chose \texttt{\DIFdelbegin \DIFdel{Seq-Gen}\DIFdelend \DIFaddbegin \DIFadd{INDELible}\DIFaddend } because it is used by \texttt{\DIFdelbegin \DIFdel{TreeToReads}\DIFdelend \DIFaddbegin \DIFadd{NGSphy}\DIFaddend } and because it can generate
sequences \DIFdelbegin \DIFdel{based on gene trees.
It is relatively simple and written in C, so it should provide a conservative
estimate of how well }\texttt{\DIFdel{jackalope}} %DIFAUXCMD
\DIFdel{performs against similar programs.
}\DIFdelend \DIFaddbegin \DIFadd{with both substitutions and indels based on phylogenetic trees.
}\DIFaddend I tested each program by having them simulate a 2 Mb \DIFdelbegin \DIFdel{, }\DIFdelend \DIFaddbegin \DIFadd{and }\DIFaddend 20 Mb \DIFdelbegin \DIFdel{, and 200 Mb }\DIFdelend genome
split evenly among 20 \DIFdelbegin \DIFdel{sequences}\DIFdelend \DIFaddbegin \DIFadd{chromosomes}\DIFaddend , then generate 8 variants from that genome using
\DIFdelbegin \DIFdel{gene }\DIFdelend \DIFaddbegin \DIFadd{phylogenetic }\DIFaddend trees produced by the \DIFdelbegin \DIFdel{following }\texttt{\DIFdel{scrm}} %DIFAUXCMD
\DIFdel{command :
}%DIFDELCMD < 

%DIFDELCMD < \begin{Shaded}
%DIFDELCMD < \begin{Highlighting}[]
%DIFDELCMD < %%%
\DIFdel{\ExtensionTok{scrm}\NormalTok{ 8 20 -T -seed 1829812441 253160851 137767610 }\OperatorTok{>}\NormalTok{ scrm.tree}
}%DIFDELCMD < \end{Highlighting}
%DIFDELCMD < \end{Shaded}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{I }\DIFdelend \DIFaddbegin \DIFadd{command line version of }\texttt{\DIFadd{scrm}}\DIFadd{.
I ran separate simulations with the trees scaled to maximum tree depths of 0.1, 0.01,
and 0.001.
I }\DIFaddend used the HKY85 substitution model with \DIFdelbegin \DIFdel{an overall mutation rate of 0.001 }\DIFdelend \DIFaddbegin \DIFadd{a transition rate of 2.5 }\DIFaddend substitutions per
site per unit of time (where one unit of time is equal to a branch length of 1)\DIFdelbegin \DIFdel{.
I did not include indels because }\texttt{\DIFdel{Seq-Gen}} %DIFAUXCMD
\DIFdel{cannot simulate them.
Among-site variability was generated using a }\DIFdelend \DIFaddbegin \DIFadd{,
and a transversion rate of 1.
Relative frequencies of nucleotides were 0.4, 0.3, 0.2, and 0.1 for T, C, A, and G,
respectively.
For among-site variability in substitution rates, I used
(1) a discrete }\DIFaddend Gamma distribution with \DIFaddbegin \DIFadd{10 categories and }\DIFaddend a shape parameter of \DIFdelbegin \DIFdel{1; in }\texttt{\DIFdel{jackalope}}%DIFAUXCMD
\DIFdel{, the region size was 100 bp.
}\DIFdelend \DIFaddbegin \DIFadd{0.5, and
(2) an invariant-site rate of 0.25.
The total rate for insertions was 0.1 per site per unit time,
with relative rates derived from a Lavalette distribution with \(L = 541\) and \(a = 1.7\).
Deletions had the same total and relative rates.
}\DIFaddend Output was written to \DIFdelbegin \DIFdel{either PHYLIP format (for }\texttt{\DIFdel{Seq-Gen}}%DIFAUXCMD
\DIFdel{) or
FASTA files(for }\texttt{\DIFdel{jackalope}}%DIFAUXCMD
\DIFdel{)}\DIFdelend \DIFaddbegin \DIFadd{FASTA files}\DIFaddend .
\texttt{jackalope} was tested for 1 and 4 threads, but
\texttt{\DIFdelbegin \DIFdel{Seq-Gen}\DIFdelend \DIFaddbegin \DIFadd{INDELible}\DIFaddend } for only 1 because it cannot use multiple threads.

\texttt{jackalope} consistently \DIFdelbegin \DIFdel{outperforms }\DIFdelend \DIFaddbegin \DIFadd{outperformed }\DIFaddend \texttt{\DIFdelbegin \DIFdel{Seq-Gen}\DIFdelend \DIFaddbegin \DIFadd{INDELible}\DIFaddend } in terms of speed, \DIFdelbegin \DIFdel{and the effect
of using 4 threads is most pronounced at the larger genome sizes
}\DIFdelend \DIFaddbegin \DIFadd{although this
difference was less drastic when there were more mutations: for the larger genome size
and greater maximum tree depth }\DIFaddend (Figure \ref{fig:variants-perf-test-plot}).
This \DIFdelbegin \DIFdel{makes sense, since the overhead associated with multithreading becomes more worthwhile when there are more tasks to accomplish}\DIFdelend \DIFaddbegin \DIFadd{is somewhat offset by the fact that using more threads has a greater effect at
larger genome sizes}\DIFaddend .
Memory usage \DIFdelbegin \DIFdel{is }\DIFdelend \DIFaddbegin \DIFadd{was }\DIFaddend greater in \texttt{jackalope} \DIFdelbegin \DIFdel{at low genome sizes }\DIFdelend \DIFaddbegin \DIFadd{than }\texttt{\DIFadd{INDELible}}\DIFadd{.
With low mutation numbers, this difference was }\DIFaddend due to the overhead associated
with loading R\DIFdelbegin \DIFdel{:
}\DIFdelend \DIFaddbegin \DIFadd{.
For a 2 Mb genome and 0.001 max tree depth, }\texttt{\DIFadd{jackalope}} \DIFadd{used
94.3 MB memory,
while }\texttt{\DIFadd{INDELible}} \DIFadd{used 20.7 MB.
}\DIFaddend An R script that simply printed an empty string used \(\sim 55\) MB memory,
and another that only loaded \texttt{jackalope} used \(\sim 70\) MB.
\DIFdelbegin \DIFdel{At the }\DIFdelend \DIFaddbegin \DIFadd{With greater numbers of mutations, the difference was due to }\texttt{\DIFadd{jackalope}} \DIFadd{using
64-bit integers to store positions instead of 32-bit.
In the most extreme example (}\DIFaddend 20 Mb genome \DIFdelbegin \DIFdel{size, however, }\DIFdelend \DIFaddbegin \DIFadd{and 0.1 max tree depth), }\DIFaddend \texttt{jackalope} used
\DIFdelbegin \DIFdel{slightly less memory than }\texttt{\DIFdel{Seq-Gen}}%DIFAUXCMD
\DIFdel{,
and for a 200 Mb genome size, }\texttt{\DIFdel{jackalope}} %DIFAUXCMD
\DIFdelend \DIFaddbegin \DIFadd{768.3 MB memory and }\texttt{\DIFadd{INDELible}} \DIFaddend used
\DIFdelbegin \DIFdel{\(38 \%\) less memory}\DIFdelend \DIFaddbegin \DIFadd{429.3 MB.
I chose to use 64-bit integers in }\texttt{\DIFadd{jackalope}} \DIFadd{to allow chromosomes to exceed
2 Gb in length.
Memory usage increased by only 31.7kB when
}\texttt{\DIFadd{jackalope}} \DIFadd{used 4 threads}\DIFaddend .

\hypertarget{generating-sequencing-reads}{%
\subsubsection{Generating sequencing reads}\label{generating-sequencing-reads}}

For both Illumina and PacBio read generation in \texttt{jackalope}, I compared their performance
to the programs they were based on:
\texttt{ART} (\DIFaddbegin \DIFadd{version 2.5.8; }\DIFaddend Huang et al., 2011) and
\texttt{SimLoRD} (\DIFaddbegin \DIFadd{version 1.0.3; }\DIFaddend Stöcker et al., 2016), respectively.
\texttt{ART} has the added benefit of being used for both \texttt{NSGphy} and \texttt{TreesToReads}.
All these tests consisted of reading a 2 Mb genome from a FASTA file, simulating
reads, and writing them to FASTQ files.
No ALN or SAM/BAM files were written by any of the programs.
For Illumina sequencing, I generated \(2 \times 150\) reads from the HiSeq 2500 platform,
and tests were conducted for 100 thousand, 1 million, and 10 million total reads.
For PacBio sequencing, I generated reads using default parameters,
and tests were conducted for 1, 10, and 100 thousand total reads.
\texttt{jackalope} was tested for 1 and 4 threads, but
neither \texttt{ART} nor \texttt{SimLoRD} allowed the use of more than 1.

Again, \texttt{jackalope} outperformed both \texttt{ART} and \texttt{SimLoRD} in terms of speed
(Figure \ref{fig:sequencing-perf-test-plot}\DIFdelbegin \DIFdel{).
Using multiple threads }\DIFdelend \DIFaddbegin \DIFadd{A--B).
Multithreading }\DIFaddend was most useful when many reads were generated:
For at least 10M Illumina reads and at least \DIFdelbegin \DIFdel{1M }\DIFdelend \DIFaddbegin \DIFadd{100k }\DIFaddend PacBio reads, using 4 threads
reduced the elapsed time for jobs by \(\sim 50 \%\).
All three programs used very little memory\DIFdelbegin \DIFdel{, and
the amount did not depend strongly
on the number of reads generated.
}\texttt{\DIFdel{ART}} %DIFAUXCMD
\DIFdel{used 9.4 MBand }\DIFdelend \DIFaddbegin \DIFadd{:
Average memory usage was 9.3 MB for }\texttt{\DIFadd{ART}} \DIFadd{and
60.9 for }\DIFaddend \texttt{SimLoRD}\DIFdelbegin \DIFdel{used 54.0--61.2 MB}\DIFdelend .
\texttt{jackalope} \DIFdelbegin \DIFdel{used 76.5--79.4 MB }\DIFdelend \DIFaddbegin \DIFadd{averaged 73.7 MB for Illumina reads and
102.0 for PacBio reads.
Memory usage increased only slightly with increasing read numbers in }\texttt{\DIFadd{jackalope}}\DIFadd{:
It increased 0.6 MB across the range of Illumina reads and 3 MB across the PacBio range.
Over the Illumina range, memory usage by }\texttt{\DIFadd{ART}} \DIFadd{slightly decreased
(-0.05 MB).
}\texttt{\DIFadd{SimLoRD}} \DIFadd{memory usage increased over its range
(8 MB).
Using 4 threads in }\texttt{\DIFadd{jackalope}} \DIFadd{had little effect on memory usage
(3.2 MB extra }\DIFaddend for Illumina reads\DIFaddbegin \DIFadd{,
6.3 MB extra for PacBio).
}

\hypertarget{pipeline-from-reference-to-reads}{%
\subsubsection{Pipeline from reference to reads}\label{pipeline-from-reference-to-reads}}

\DIFadd{I lastly compared the performance of }\texttt{\DIFadd{jackalope}} \DIFadd{to that of }\texttt{\DIFadd{NGSphy}}
\DIFadd{(version 1.0.13; Escalona et al., 2018)
for the following pipeline:
generating a reference genome, simulating variants from gene trees, and
producing Illumina reads from the variants.
For the first two steps, I used the same parameters as I did for the comparison to
}\texttt{\DIFadd{INDELible}}\DIFadd{.
I also used the same parameters for the last step, except that the haploid variants
were used for sequencing reads instead of a reference genome.
I generated 8 variants with 20 chromosomes by creating a phylogenetic tree with
160 tips and assigning every 20 tips to the same individual using the naming
scheme necessary for }\texttt{\DIFadd{NGSphy}}\DIFadd{.
I divided the larger tree by chromosome and passed these separate trees to
}\texttt{\DIFadd{jackalope}} \DIFadd{by writing them to a }\texttt{\DIFadd{ms}}\DIFadd{-style output file.
Output was written to both FASTA and FASTQ files.
Both }\texttt{\DIFadd{NGSphy}} \DIFaddend and \DIFdelbegin \DIFdel{101.3--110.1 MB for PacBio reads.
}\DIFdelend \DIFaddbegin \texttt{\DIFadd{jackalope}} \DIFadd{were tested using both 1 and 4 threads.
}\DIFaddend 

\DIFaddbegin \texttt{\DIFadd{jackalope}} \DIFadd{was faster than }\texttt{\DIFadd{NGSphy}} \DIFadd{for all the performance tests
(Figure \ref{fig:pipeline-perf-test-plot}A).
The speed advantage of }\texttt{\DIFadd{jackalope}} \DIFadd{was greatest
at low mutation densities and when the number of reads generated was higher.
The effect of multithreading in }\texttt{\DIFadd{jackalope}} \DIFadd{increased with increasing read numbers,
genome sizes, or maximum tree depth.
These results were both consistent with those from the separate performance tests with
}\texttt{\DIFadd{INDELible}} \DIFadd{and }\texttt{\DIFadd{ART}}\DIFadd{, the programs that make up the }\texttt{\DIFadd{NGSphy}} \DIFadd{pipeline.
Attempts at multithreading in }\texttt{\DIFadd{NGSphy}} \DIFadd{had no effect on its performance:
Calling for 4 threads resulted in 0.7 \% more
elapsed time and never caused total CPU time to exceed elapsed time.
Thus multithreading in }\texttt{\DIFadd{NGSphy}} \DIFadd{seems to take place in another portion of the pipeline
that is not apparent from its documentation.
For this reason, I have grouped results from both the 1- and 4-thread calls to }\texttt{\DIFadd{NGSphy}}
\DIFadd{together in Figure \ref{fig:pipeline-perf-test-plot} and in the text below.
}

\DIFadd{Throughout the tests, }\texttt{\DIFadd{NGSphy}} \DIFadd{used significantly more memory than }\texttt{\DIFadd{jackalope}}
\DIFadd{(Figure \ref{fig:pipeline-perf-test-plot}B).
This was particularly true for the 20 Mb genome at maximum tree depths of 0.001 and 0.01,
where }\texttt{\DIFadd{NGSphy}} \DIFadd{consumed
892 \%
and
947 \%
more memory, respectively.
The number of reads had little effect on memory usage: When generating 10M versus
100k reads, }\texttt{\DIFadd{jackalope}} \DIFadd{consumed 0.02 \%
less memory, and }\texttt{\DIFadd{NGSphy}} \DIFadd{used 0.6 \%
less.
}

\DIFaddend \hypertarget{example-usage}{%
\subsection{Example usage}\label{example-usage}}

This section provides brief examples of how \texttt{jackalope} can be used
to generate sequencing data that can inform some common sampling decisions for HTS
studies.
These examples only show how to produce the output from \texttt{jackalope}, as
a review of different bioinformatic programs that would process these data
is beyond the scope of the present manuscript.

For an example reference assembly, I used the \emph{Drosophila melanogaster} assembly
(version 6.27) downloaded from \texttt{flybase.org} (Thurmond et al., 2018).
After downloading, I read the compressed FASTA file, filtered out
scaffolds by using a size threshold, and manually removed the sex chromosomes
using the following code:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ref <-}\StringTok{ }\KeywordTok{read_fasta}\NormalTok{(}\StringTok{"dmel-6.27.fasta.gz"}\NormalTok{, }\DataTypeTok{cut_names =} \OtherTok{TRUE}\NormalTok{)}
\NormalTok{ref}\OperatorTok{$}\DIFdelbegin \DIFdel{\KeywordTok{filter_seqs}}\DIFdelend \DIFaddbegin \DIFadd{\KeywordTok{filter_chroms}}\DIFaddend \NormalTok{(}\FloatTok{1e6}\NormalTok{, }\DataTypeTok{method =} \StringTok{"size"}\NormalTok{)}
\NormalTok{ref}\OperatorTok{$}\DIFdelbegin \DIFdel{\KeywordTok{rm_seqs}}\DIFdelend \DIFaddbegin \DIFadd{\KeywordTok{rm_chroms}}\DIFaddend \NormalTok{(}\KeywordTok{c}\NormalTok{(}\StringTok{"X"}\NormalTok{, }\StringTok{"Y"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\DIFdelbegin \DIFdel{This resulted in the following }\texttt{\DIFdel{ref\_genome}} %DIFAUXCMD
\DIFdel{object:
}%DIFDELCMD < 

%DIFDELCMD < \begin{verbatim}%DIFDELCMD < 
%DIFDELCMD < ## < Set of 5 sequences >
%DIFDELCMD < ## # Total size: 110,338,337 bp
%DIFDELCMD < ##   name                          sequence                             length
%DIFDELCMD < ## 3R         ACGGGACCGAGTATAGTACCAGTAC...CGTGACTGTTCGCATTCTAGGAATTC  32079331
%DIFDELCMD < ## 3L         TAGGGAGAAATATGATCGCGTATGC...TTCTATGTTATTCCATGTTATTCTAT  28110227
%DIFDELCMD < ## 2R         CTCAAGATACCTTCTACAGATTATT...TGGAGGTACGGAAACAGAATGAATTC  25286936
%DIFDELCMD < ## 2L         CGACAATGCACGACAGAGGAAGCAG...AACAGAGAACAGAGAACAGAGAAGAG  23513712
%DIFDELCMD < ## 4          TTATTATATTATTATATTATTATAT...GCAGCCGTCGATTTGAGATATATGAA   1348131
%DIFDELCMD < \end{verbatim}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{For molecular-evolution }\DIFdelend \DIFaddbegin \DIFadd{For mutation-type }\DIFaddend information, I used the \DIFdelbegin \DIFdel{TN93 }\DIFdelend \DIFaddbegin \DIFadd{JC69 substitution }\DIFaddend model,
an insertion rate of \texttt{2e-5} for lengths from 1 to 10,
and
a deletion rate of \texttt{1e-5} for lengths from 1 to 40.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sub <-}\StringTok{ }\DIFdelbegin \DIFdel{\KeywordTok{sub_TN93}\NormalTok{(}\DataTypeTok{pi_tcag =} \KeywordTok{c}\NormalTok{(}\FloatTok{0.1}\NormalTok{, }\FloatTok{0.2}\NormalTok{, }\FloatTok{0.3}\NormalTok{, }\FloatTok{0.4}\NormalTok{),}
                \DataTypeTok{alpha_1 =} \FloatTok{0.0001}\NormalTok{, }\DataTypeTok{alpha_2 =} \FloatTok{0.0002}\NormalTok{,}
                \DataTypeTok{beta =} \FloatTok{0.00015}}\DIFdelend \DIFaddbegin \DIFadd{\KeywordTok{sub_JC69}\NormalTok{(}\DataTypeTok{lambda =} \FloatTok{1e-4}}\DIFaddend \NormalTok{)}
\NormalTok{ins <-}\StringTok{ }\KeywordTok{indels}\NormalTok{(}\DataTypeTok{rate =} \FloatTok{2e-5}\NormalTok{, }\DataTypeTok{max_length =} \DecValTok{10}\NormalTok{)}
\NormalTok{del <-}\StringTok{ }\KeywordTok{indels}\NormalTok{(}\DataTypeTok{rate =} \FloatTok{1e-5}\NormalTok{, }\DataTypeTok{max_length =} \DecValTok{40}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{assembling-a-genome}{%
\subsubsection{Assembling a genome}\label{assembling-a-genome}}

The example here produces FASTQ files from the known \DIFaddbegin \emph{\DIFadd{D. melanogaster}} \DIFaddend reference
assembly that could test strategies for how to assemble a similar genome using HTS data.
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{The first strategy is to use only PacBio sequencing.
The PacBio Sequel system produces up to 500,
000 readsper
Single Molecule, Real-Time (SMRT) cell, so you could
run }\DIFdelend \DIFaddbegin \DIFadd{The strategies the example code below could test would be
(1) a PacBio-only assembly,
(2) a PacBio-only assembly followed by polishing using Illumina reads, and
(3) an explicit hybrid assembly.
}

\DIFadd{To generate data to test the first, PacBio-only strategy, I ran
}\DIFaddend the following for two \DIFdelbegin \DIFdel{cells }\DIFdelend \DIFaddbegin \DIFadd{Single Molecule, Real-Time (SMRT) cells from the PacBio Sequel
system }\DIFaddend (with the file \texttt{pacbio\_R1.fq} as output):

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{pacbio}\NormalTok{(ref, }\DataTypeTok{out_prefix =} \StringTok{"pacbio"}\NormalTok{, }\DataTypeTok{n_reads =} \DecValTok{2} \OperatorTok{*}\StringTok{ }\FloatTok{500e3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\DIFdelbegin \DIFdel{An alternative, hybrid strategy uses
}\DIFdelend \DIFaddbegin \DIFadd{For either of the last two strategies, I used
}\DIFaddend 1 SMRT cell of PacBio sequencing and
1 lane (\(\sim 500\) million reads) of \(2 \times 100\)bp Illumina
sequencing on the HiSeq 2500 system (the default Illumina system in \texttt{jackalope}):

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{pacbio}\NormalTok{(ref, }\DataTypeTok{out_prefix =} \StringTok{"pacbio"}\NormalTok{, }\DataTypeTok{n_reads =} \FloatTok{500e3}\NormalTok{)}
\KeywordTok{illumina}\NormalTok{(ref, }\DataTypeTok{out_prefix =} \StringTok{"illumina"}\NormalTok{, }\DataTypeTok{n_reads =} \FloatTok{500e6}\NormalTok{, }\DataTypeTok{paired =} \OtherTok{TRUE}\NormalTok{,}
         \DataTypeTok{read_length =} \DecValTok{100}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\DIFdelbegin \DIFdel{The last strategycombines 1 lane of \(2 \times 100\)bp Illumina HiSeq 2500 sequencing with 1 flow cell of \(2 \times 250\)bp mate-pair sequencing on an Illumina MiSeq v3.
The mate-pair sequencing uses longer fragments (defaults are mean of 400 and standard deviation of 100) to better cover highly
repetitive regions}\DIFdelend \DIFaddbegin \DIFadd{These data could then be used to compare genome assembly performance between
the strategies above, or between programs within a given strategy.
Extensions of these tests include adjusting sequencing depth, sequencing platform,
and error rates}\DIFaddend .

\DIFdelbegin %DIFDELCMD < \begin{Shaded}
%DIFDELCMD < \begin{Highlighting}[]
%DIFDELCMD < %%%
\DIFdel{\KeywordTok{illumina}\NormalTok{(ref, }\DataTypeTok{out_prefix =} \StringTok{"ill_pe"}\NormalTok{, }\DataTypeTok{n_reads =} \FloatTok{500e6}\NormalTok{, }\DataTypeTok{paired =} \OtherTok{TRUE}\NormalTok{,}
         \DataTypeTok{read_length =} \DecValTok{100}\NormalTok{)}
\KeywordTok{illumina}\NormalTok{(ref, }\DataTypeTok{out_prefix =} \StringTok{"ill_mp"}\NormalTok{, }\DataTypeTok{seq_sys =} \StringTok{"MSv3"}\NormalTok{,}
         \DataTypeTok{read_length =} \DecValTok{250}\NormalTok{, }\DataTypeTok{n_reads =} \FloatTok{50e6}\NormalTok{, }\DataTypeTok{matepair =} \OtherTok{TRUE}\NormalTok{, }
         \DataTypeTok{frag_mean =} \DecValTok{3000}\NormalTok{, }\DataTypeTok{frag_sd =} \DecValTok{500}\NormalTok{)}
}%DIFDELCMD < \end{Highlighting}
%DIFDELCMD < \end{Shaded}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend \hypertarget{estimating-divergence-between-populations}{%
\subsubsection{Estimating divergence between populations}\label{estimating-divergence-between-populations}}

Here, I will demonstrate how to generate \DIFaddbegin \DIFadd{pooled }\DIFaddend population-genomic data \DIFdelbegin \DIFdel{of a type that might
}\DIFdelend \DIFaddbegin \DIFadd{(Pool-seq)
that could }\DIFaddend be used to estimate the divergence between two populations.
\DIFdelbegin \DIFdel{I first use }\DIFdelend \DIFaddbegin \DIFadd{Before starting }\texttt{\DIFadd{jackalope}}\DIFadd{, I used }\DIFaddend the \texttt{scrm} package (Staab et al., 2015)
to conduct coalescent simulations that will generate segregating sites for 40
haploid variants from the reference genome \DIFaddbegin \DIFadd{(object }\texttt{\DIFadd{ssites}}\DIFadd{)}\DIFaddend .
Twenty of the variants are from one population, twenty from another.
The symmetrical migration rate is 100 individuals per generation\DIFdelbegin \DIFdel{.
To generate many mutations,
I set }\DIFdelend \DIFaddbegin \DIFadd{,
and }\DIFaddend the population-scaled mutation rate (\(\theta = 4 N_0 \mu\)) \DIFdelbegin \DIFdel{to }\DIFdelend \DIFaddbegin \DIFadd{is 10.
The }\DIFaddend \texttt{\DIFdelbegin \DIFdel{1000}%DIFDELCMD < } %%%
\DIFdel{for this example}\DIFdelend \DIFaddbegin \DIFadd{scrm}} \DIFadd{command for this is }\texttt{\DIFadd{40\ 5\ -t\ 10\ -I\ 2\ 20\ 20\ 100}}\DIFaddend .

\DIFdelbegin %DIFDELCMD < \begin{Shaded}
%DIFDELCMD < \begin{Highlighting}[]
%DIFDELCMD < %%%
\DIFdel{\NormalTok{ssites <-}\StringTok{ }\KeywordTok{scrm}\NormalTok{(}\KeywordTok{paste}\NormalTok{(}\StringTok{"40"}\NormalTok{, ref}\OperatorTok{$}\KeywordTok{n_seqs}\NormalTok{(), }\StringTok{"-t 1000 -I 2 20 20 100"}\NormalTok{))}
}%DIFDELCMD < \end{Highlighting}
%DIFDELCMD < \end{Shaded}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend Using the previously created objects for \DIFdelbegin \DIFdel{molecular evolution information (}\texttt{\DIFdel{sub}}%DIFAUXCMD
\DIFdel{,
}\texttt{\DIFdel{ins}}%DIFAUXCMD
\DIFdel{, and }\DIFdelend \DIFaddbegin \DIFadd{mutation-type information and the
}\DIFaddend \texttt{\DIFdelbegin \DIFdel{del}\DIFdelend \DIFaddbegin \DIFadd{vars\_ssites}\DIFaddend } \DIFdelbegin \DIFdel{),
I create }\DIFdelend \DIFaddbegin \DIFadd{function that checks and organizes the coalescent output object,
I created }\DIFaddend variants from the reference genome:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{vars <-}\StringTok{ }\KeywordTok{create_variants}\NormalTok{(ref, }\KeywordTok{vars_ssites}\NormalTok{(ssites), sub, ins, del)}
\end{Highlighting}
\end{Shaded}

\DIFdelbegin \DIFdel{This results in the following set of variants:
}%DIFDELCMD < 

%DIFDELCMD < \begin{verbatim}%DIFDELCMD < 
%DIFDELCMD < ##                            << Variants object >>
%DIFDELCMD < ## # Variants: 40
%DIFDELCMD < ## # Mutations: 440,203
%DIFDELCMD < ##
%DIFDELCMD < ##                         << Reference genome info: >>
%DIFDELCMD < ## < Set of 5 sequences >
%DIFDELCMD < ## # Total size: 110,338,337 bp
%DIFDELCMD < ##   name                          sequence                             length
%DIFDELCMD < ## 3R         ACGGGACCGAGTATAGTACCAGTAC...CGTGACTGTTCGCATTCTAGGAATTC  32079331
%DIFDELCMD < ## 3L         TAGGGAGAAATATGATCGCGTATGC...TTCTATGTTATTCCATGTTATTCTAT  28110227
%DIFDELCMD < ## 2R         CTCAAGATACCTTCTACAGATTATT...TGGAGGTACGGAAACAGAATGAATTC  25286936
%DIFDELCMD < ## 2L         CGACAATGCACGACAGAGGAAGCAG...AACAGAGAACAGAGAACAGAGAAGAG  23513712
%DIFDELCMD < ## 4          TTATTATATTATTATATTATTATAT...GCAGCCGTCGATTTGAGATATATGAA   1348131
%DIFDELCMD < \end{verbatim}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{For a file of true divergences from the reference genome, the }\texttt{\DIFdel{write\_vcf}} %DIFAUXCMD
\DIFdel{function
writes the }\texttt{\DIFdel{variants}} %DIFAUXCMD
\DIFdel{object to VCF file }\texttt{\DIFdel{variants.vcf}}%DIFAUXCMD
\DIFdel{:
}%DIFDELCMD < 

%DIFDELCMD < \begin{Shaded}
%DIFDELCMD < \begin{Highlighting}[]
%DIFDELCMD < %%%
\DIFdel{\KeywordTok{write_vcf}\NormalTok{(vars, }\StringTok{"variants"}\NormalTok{)}
}%DIFDELCMD < \end{Highlighting}
%DIFDELCMD < \end{Shaded}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend Lastly, I \DIFdelbegin \DIFdel{simulate }\DIFdelend \DIFaddbegin \DIFadd{simulated }\DIFaddend 1 lane of \(2 \times 100\)bp Illumina HiSeq 2500 sequencing.
In this case, \DIFdelbegin \DIFdel{individuals within a population are pooled, and the population
sequences are derived from are identified by barcodes.
}%DIFDELCMD < 

%DIFDELCMD < \begin{Shaded}
%DIFDELCMD < \begin{Highlighting}[]
%DIFDELCMD < %%%
\DIFdel{\KeywordTok{illumina}\NormalTok{(vars, }\DataTypeTok{out_prefix =} \StringTok{"vars_illumina"}\NormalTok{, }\DataTypeTok{n_reads =} \FloatTok{500e6}\NormalTok{, }\DataTypeTok{paired =} \OtherTok{TRUE}\NormalTok{,}
         \DataTypeTok{read_length =} \DecValTok{100}\NormalTok{, }\DataTypeTok{barcodes =} \KeywordTok{c}\NormalTok{(}\KeywordTok{rep}\NormalTok{(}\StringTok{"AACCGCGG"}\NormalTok{, }\DecValTok{20}\NormalTok{), }
                                         \KeywordTok{rep}\NormalTok{(}\StringTok{"GGTTATAA"}\NormalTok{, }\DecValTok{20}\NormalTok{)))}
}%DIFDELCMD < \end{Highlighting}
%DIFDELCMD < \end{Shaded}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{The below example instead has }\DIFdelend each individual variant's reads \DIFaddbegin \DIFadd{were }\DIFaddend output to separate FASTQ files\DIFdelbegin \DIFdel{:
}\DIFdelend \DIFaddbegin \DIFadd{.
I also wrote the true variant information to a VCF file.
}\DIFaddend 

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{illumina}\NormalTok{(vars, }\DataTypeTok{out_prefix =} \StringTok{"vars_illumina"}\NormalTok{, }\DataTypeTok{n_reads =} \FloatTok{500e6}\NormalTok{, }\DataTypeTok{paired =} \OtherTok{TRUE}\NormalTok{,}
         \DataTypeTok{read_length =} \DecValTok{100}\NormalTok{, }\DataTypeTok{sep_files =} \OtherTok{TRUE}\NormalTok{)}
\DIFaddbegin \DIFadd{\KeywordTok{write_vcf}\NormalTok{(vars, }\StringTok{"variants"}\NormalTok{)}
}\DIFaddend \end{Highlighting}
\end{Shaded}

\DIFaddbegin \DIFadd{The \(F_{ST}\) calculated from the resulting VCF file could then be compared to output
from various programs to inform which works best in a particular case.
For uncertain population parameters (e.g., migration rates), output from
multiple calls to }\texttt{\DIFadd{scrm}} \DIFadd{varying the parameter of interest could be input to
the }\texttt{\DIFadd{jackalope}} \DIFadd{pipeline above to identify the conditions under which one program
might have an advantage over another.
}

\DIFaddend \hypertarget{constructing-a-phylogeny}{%
\subsubsection{Constructing a phylogeny}\label{constructing-a-phylogeny}}

\hypertarget{from-one-phylogenetic-tree}{%
\paragraph{From one phylogenetic tree}\label{from-one-phylogenetic-tree}}

This section shows how \texttt{jackalope} can generate variants from a phylogeny, then
simulate sequencing data from those variants to test phylogeny reconstruction methods.
First, I \DIFdelbegin \DIFdel{generated a random coalescent tree }\DIFdelend \DIFaddbegin \DIFadd{simulated a coalescent phylogeny }\DIFaddend of 10 species \DIFdelbegin \DIFdel{using the }\DIFdelend \DIFaddbegin \DIFadd{(object }\texttt{\DIFadd{tree}}\DIFadd{) using
}\DIFaddend \texttt{ape\DIFaddbegin \DIFadd{::rcoal(10)}\DIFaddend } \DIFdelbegin \DIFdel{package
}\DIFdelend (Paradis \& Schliep, 2018).
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < \begin{Shaded}
%DIFDELCMD < \begin{Highlighting}[]
%DIFDELCMD < %%%
\DIFdel{\NormalTok{tree <-}\StringTok{ }\NormalTok{ape}\OperatorTok{::}\KeywordTok{rcoal}\NormalTok{(}\DecValTok{10}\NormalTok{)}
}%DIFDELCMD < \end{Highlighting}
%DIFDELCMD < \end{Shaded}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{Then I input that to the }\DIFdelend \DIFaddbegin \DIFadd{Function }\texttt{\DIFadd{vars\_phylo}} \DIFadd{organizes and checks the }\DIFaddend \texttt{\DIFdelbegin \DIFdel{create\_variants}\DIFdelend \DIFaddbegin \DIFadd{tree}\DIFaddend } \DIFdelbegin \DIFdel{function alongside the molecular evolution
information .
}\DIFdelend \DIFaddbegin \DIFadd{object, and
including it with the mutation-type information allowed me to create variants based
on this phylogeny:
}\DIFaddend 

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{vars <-}\StringTok{ }\KeywordTok{create_variants}\NormalTok{(ref, }\KeywordTok{vars_phylo}\NormalTok{(tree), sub, ins, del)}
\end{Highlighting}
\end{Shaded}

\DIFdelbegin \DIFdel{This results in the following }\texttt{\DIFdel{variants}} %DIFAUXCMD
\DIFdel{object:
}%DIFDELCMD < 

%DIFDELCMD < \begin{verbatim}%DIFDELCMD < 
%DIFDELCMD < ##                            << Variants object >>
%DIFDELCMD < ## # Variants: 10
%DIFDELCMD < ## # Mutations: 301,710
%DIFDELCMD < ##
%DIFDELCMD < ##                         << Reference genome info: >>
%DIFDELCMD < ## < Set of 5 sequences >
%DIFDELCMD < ## # Total size: 110,338,337 bp
%DIFDELCMD < ##   name                          sequence                             length
%DIFDELCMD < ## 3R         ACGGGACCGAGTATAGTACCAGTAC...CGTGACTGTTCGCATTCTAGGAATTC  32079331
%DIFDELCMD < ## 3L         TAGGGAGAAATATGATCGCGTATGC...TTCTATGTTATTCCATGTTATTCTAT  28110227
%DIFDELCMD < ## 2R         CTCAAGATACCTTCTACAGATTATT...TGGAGGTACGGAAACAGAATGAATTC  25286936
%DIFDELCMD < ## 2L         CGACAATGCACGACAGAGGAAGCAG...AACAGAGAACAGAGAACAGAGAAGAG  23513712
%DIFDELCMD < ## 4          TTATTATATTATTATATTATTATAT...GCAGCCGTCGATTTGAGATATATGAA   1348131
%DIFDELCMD < \end{verbatim}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{Now I can generate }\DIFdelend \DIFaddbegin \DIFadd{I then generated }\DIFaddend data for 1 flow cell of \DIFdelbegin \DIFdel{\(2 \times 250\)}\DIFdelend \DIFaddbegin \DIFadd{\(1 \times 250\)}\DIFaddend bp sequencing
on an Illumina MiSeq v3\DIFdelbegin \DIFdel{.
}\DIFdelend \DIFaddbegin \DIFadd{, where }\texttt{\DIFadd{variant\_barcodes}} \DIFadd{is a character string that specifies
the barcodes for each variant.
I also wrote the true phylogenetic tree to a NEWICK file.
}\DIFaddend 

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{illumina}\NormalTok{(vars, }\DataTypeTok{out_prefix =} \StringTok{"phylo_tree"}\NormalTok{, }\DataTypeTok{seq_sys =} \StringTok{"MSv3"}\NormalTok{,}
         \DIFaddbegin \DIFadd{\DataTypeTok{paired =} \OtherTok{FALSE}\NormalTok{, }}\DIFaddend \DataTypeTok{read_length =} \DecValTok{250}\NormalTok{, }\DataTypeTok{n_reads =} \FloatTok{50e6}\DIFaddbegin \DIFadd{\NormalTok{,}
         \DataTypeTok{barcodes =}\NormalTok{ variant_barcodes)}
\NormalTok{ape}\OperatorTok{::}\KeywordTok{write.tree}\NormalTok{(tree, }\StringTok{"true.tree"}}\DIFaddend \NormalTok{)}
\end{Highlighting}
\end{Shaded}

\DIFaddbegin \DIFadd{The true phylogenetic tree would then be compared to the final tree output from
the program(s) the user chooses to test.
}

\DIFaddend \hypertarget{from-gene-trees}{%
\paragraph{From gene trees}\label{from-gene-trees}}

Similar to the section above, the ultimate goal here is to test phylogeny
reconstruction methods.
The difference in this section is that instead of using a single, straightforward
phylogeny, I use multiple gene trees per \DIFdelbegin \DIFdel{sequence}\DIFdelend \DIFaddbegin \DIFadd{chromosome}\DIFaddend .
In the species used in these simulations, species 3 diverged from 1 and 2 at \(t = 1.0\),
where \(t\) indicates time into the past and is in units of \(4 N_0\) generations.
Species 1 and 2 diverged at \(t = 0.5\).
I assume a recombination rate of \(1 / (4 N_0)\) recombination events per \DIFdelbegin \DIFdel{sequence
}\DIFdelend \DIFaddbegin \DIFadd{chromosome
}\DIFaddend per generation.
\DIFdelbegin \DIFdel{Because each sequence is a different length and the length is required for including
a recombination rate, I had to run }\texttt{\DIFdel{scrm}} %DIFAUXCMD
\DIFdel{separately for each sequence.
}\DIFdelend There are 4 diploid individuals sampled per species.
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < \begin{Shaded}
%DIFDELCMD < \begin{Highlighting}[]
%DIFDELCMD < %%%
\DIFdel{\CommentTok{# Run scrm for one sequence size:}
\NormalTok{one_seq <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(size) \{}
\NormalTok{    sims <-}\StringTok{ }\KeywordTok{scrm}\NormalTok{(}
        \KeywordTok{paste}\NormalTok{(}\StringTok{"24 1"}\NormalTok{,}
              \CommentTok{# Output gene trees:}
              \StringTok{"-T"}\NormalTok{,}
              \CommentTok{# Recombination:}
              \StringTok{"-r 1"}\NormalTok{, size,}
              \CommentTok{# 3 species with no ongoing migration:}
              \StringTok{"-I 3"}\NormalTok{, }\KeywordTok{paste}\NormalTok{(}\KeywordTok{rep}\NormalTok{(}\StringTok{"8"}\NormalTok{, }\DecValTok{3}\NormalTok{), }\DataTypeTok{collapse =} \StringTok{" "}\NormalTok{), }\StringTok{"0"}\NormalTok{,}
              \CommentTok{# Species 3 derived from 1 (and 2) at time 1.0:}
              \StringTok{"-ej 1.0 3 1"}\NormalTok{,}
              \CommentTok{# Species 2 derived from 1 at time 0.5:}
              \StringTok{"-ej 0.5 2 1"}
\NormalTok{        ))}
    \KeywordTok{return}\NormalTok{(sims}\OperatorTok{$}\NormalTok{trees[[}\DecValTok{1}\NormalTok{]])}
\NormalTok{\}}
\CommentTok{# For all sequences:}
\NormalTok{gtrees <-}\StringTok{ }\KeywordTok{list}\NormalTok{(}\DataTypeTok{trees =} \KeywordTok{lapply}\NormalTok{(ref}\OperatorTok{$}\KeywordTok{sizes}\NormalTok{(), one_seq))}
}%DIFDELCMD < \end{Highlighting}
%DIFDELCMD < \end{Shaded}
%DIFDELCMD < 

%DIFDELCMD < \begin{Shaded}
%DIFDELCMD < \begin{Highlighting}[]
%DIFDELCMD < %%%
\DIFdel{\NormalTok{vars <-}\StringTok{ }\KeywordTok{create_variants}\NormalTok{(ref, }\KeywordTok{vars_gtrees}\NormalTok{(gtrees), sub, ins, del)}
}%DIFDELCMD < \end{Highlighting}
%DIFDELCMD < \end{Shaded}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{This results in the following }\DIFdelend \DIFaddbegin \DIFadd{I used }\DIFaddend \texttt{\DIFdelbegin \DIFdel{variants}%DIFDELCMD < } %%%
\DIFdel{object:
}%DIFDELCMD < 

%DIFDELCMD < \begin{verbatim}%DIFDELCMD < 
%DIFDELCMD < ##                            << Variants object >>
%DIFDELCMD < ## # Variants: 24
%DIFDELCMD < ## # Mutations: 613,142
%DIFDELCMD < ##
%DIFDELCMD < ##                         << Reference genome info: >>
%DIFDELCMD < ## < Set of 5 sequences >
%DIFDELCMD < ## # Total size: 110,338,337 bp
%DIFDELCMD < ##   name                          sequence                             length
%DIFDELCMD < ## 3R         ACGGGACCGAGTATAGTACCAGTAC...CGTGACTGTTCGCATTCTAGGAATTC  32079331
%DIFDELCMD < ## 3L         TAGGGAGAAATATGATCGCGTATGC...TTCTATGTTATTCCATGTTATTCTAT  28110227
%DIFDELCMD < ## 2R         CTCAAGATACCTTCTACAGATTATT...TGGAGGTACGGAAACAGAATGAATTC  25286936
%DIFDELCMD < ## 2L         CGACAATGCACGACAGAGGAAGCAG...AACAGAGAACAGAGAACAGAGAAGAG  23513712
%DIFDELCMD < ## 4          TTATTATATTATTATATTATTATAT...GCAGCCGTCGATTTGAGATATATGAA   1348131
%DIFDELCMD < \end{verbatim}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{To store mutation information by diploid sample, the }\texttt{\DIFdel{write\_vcf}} %DIFAUXCMD
\DIFdel{function writes
the }\texttt{\DIFdel{variants}} %DIFAUXCMD
\DIFdel{object to a VCF file.
It assigns haplotypes to diploid samples using a matrix for the }\DIFdelend \DIFaddbegin \DIFadd{scrm}} \DIFadd{to simulate the gene trees (command
}\DIFaddend \texttt{\DIFdelbegin \DIFdel{sample\_matrix}%DIFDELCMD < } %%%
\DIFdel{argument, where each row represents the variants for a particular
sample.
Ploidy level is not limited by }\texttt{\DIFdel{jackalope}}%DIFAUXCMD
\DIFdel{,
so polyploidy is easily simulated
by having more columns in the matrix.
The matrix for this set of diploid samples might look something like this
(only showing the first 5 rows below):
}%DIFDELCMD < 

%DIFDELCMD < \begin{verbatim}%DIFDELCMD < 
%DIFDELCMD < ##      [,1] [,2]
%DIFDELCMD < ## [1,]    1    2
%DIFDELCMD < ## [2,]    3    4
%DIFDELCMD < ## [3,]    5    6
%DIFDELCMD < ## [4,]    7    8
%DIFDELCMD < ## [5,]    9   10
%DIFDELCMD < ##          ...
%DIFDELCMD < \end{verbatim}
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdel{Using the matrix above assigns variants }\DIFdelend \DIFaddbegin \DIFadd{24\ }\DIFaddend 1\DIFdelbegin \DIFdel{and }\DIFdelend \DIFaddbegin \DIFadd{\ -T\ -r\ 1\ 1000\ -I\ 3\ 8\ 8\ 8\ 0\ -ej\ 1.0\ 3\ 1\ -ej\ 0.5\ }\DIFaddend 2\DIFdelbegin \DIFdel{to sample 1, variants 3 and 4 to
sample 2, and so on.
If the matrix is named }\DIFdelend \DIFaddbegin \DIFadd{\ 1}} \DIFadd{for a chromosome size of 1000),
producing the object }\texttt{\DIFadd{gtrees}}\DIFadd{.
As for the other variant-creation methods, function }\texttt{\DIFadd{vars\_gtrees}} \DIFadd{checks and organizes
information from the }\DIFaddend \texttt{\DIFdelbegin \DIFdel{samp\_mat}\DIFdelend \DIFaddbegin \DIFadd{gtrees}\DIFaddend } \DIFdelbegin \DIFdel{, we would use it to write the VCF file as follows:
}\DIFdelend \DIFaddbegin \DIFadd{object.
}\DIFaddend 

\begin{Shaded}
\begin{Highlighting}[]
\DIFdelbegin \DIFdel{\KeywordTok{write_vcf}\NormalTok{(vars, }\DataTypeTok{out_prefix =} \StringTok{"var_gtrees"}\NormalTok{,}
          \DataTypeTok{sample_matrix =}\NormalTok{ samp_mat)}
}\DIFdelend \DIFaddbegin \DIFadd{\NormalTok{vars <-}\StringTok{ }\KeywordTok{create_variants}\NormalTok{(ref, }\KeywordTok{vars_gtrees}\NormalTok{(gtrees), sub, ins, del)}
}\DIFaddend \end{Highlighting}
\end{Shaded}

\DIFdelbegin \DIFdel{Now I can generate }\DIFdelend \DIFaddbegin \DIFadd{Next I generated }\DIFaddend data for 1 flow cell of \(2 \times 250\)bp sequencing
on an Illumina MiSeq v3\DIFdelbegin \DIFdel{.
}\DIFdelend \DIFaddbegin \DIFadd{, with barcodes in the object }\texttt{\DIFadd{variant\_barcodes}}\DIFadd{.
I also wrote the true gene trees to a }\texttt{\DIFadd{ms}}\DIFadd{-style file.
}\DIFaddend 

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{illumina}\DIFdelbegin \DIFdel{\NormalTok{(ref, }}\DIFdelend \DIFaddbegin \DIFadd{\NormalTok{(vars, }}\DIFaddend \DataTypeTok{out_prefix =} \StringTok{"phylo_gtrees"}\NormalTok{, }\DataTypeTok{seq_sys =} \StringTok{"MSv3"}\NormalTok{,}
         \DataTypeTok{read_length =} \DecValTok{250}\NormalTok{, }\DataTypeTok{n_reads =} \FloatTok{50e6}\DIFaddbegin \DIFadd{\NormalTok{, }\DataTypeTok{paired =} \OtherTok{TRUE}\NormalTok{, }
         \DataTypeTok{barcodes =}\NormalTok{ variant_barcodes)}
\KeywordTok{write_gtrees}\NormalTok{(}\KeywordTok{vars_gtrees}\NormalTok{(gtrees), }\StringTok{"gtrees"}}\DIFaddend \NormalTok{)}
\end{Highlighting}
\end{Shaded}

\DIFaddbegin \DIFadd{Topologies of the gene trees would then be compared to final phylogenies output
from software the user is interested in testing.
Varying recombination rates or adding gene flow after separation of species would
be natural extensions of these simulations.
}

\DIFaddend \hypertarget{conclusion}{%
\subsection{Conclusion}\label{conclusion}}

\texttt{jackalope} outperforms popular stand-alone programs for phylogenomic and HTS
simulation and combines their functionalities into one cohesive package.
Although it does not provide the in-built power of \DIFdelbegin \DIFdel{full pipelines }\DIFdelend \DIFaddbegin \DIFadd{a full pipeline }\DIFaddend like \texttt{NGSphy},
simulations using \texttt{jackalope} are \DIFdelbegin \DIFdel{much }\DIFdelend simpler to implement \DIFdelbegin \DIFdel{.
Moreover, the flexibility of }\DIFdelend \DIFaddbegin \DIFadd{and more efficient computationally.
As part of a larger pipeline that includes coalescent simulations or other methods
that guide patterns of mutations, }\DIFaddend \texttt{jackalope} \DIFdelbegin \DIFdel{allows power users to manually
provide inputs in many instances where the in-built functionality may not suffice.
This package }\DIFdelend should inform research design for
projects employing HTS,
particularly those focusing on population genomics or phylogenomics.
Output from \texttt{jackalope} will help develop more specific sequencing goals
in funding applications and estimate the power of a given sequencing design.
Furthermore, \texttt{jackalope} can be used to test bioinformatic pipelines under
assumptions of much more complex evolutionary histories than most current HTS
simulation platforms allow.

\hypertarget{references}{%
\section{References}\label{references}}

\setstretch{1}

\hypertarget{refs}{}
\leavevmode\hypertarget{ref-Albers_2010}{}%
Albers, C. A., Lunter, G., MacArthur, D. G., McVean, G., Ouwehand, W. H., \& Durbin, R. (2010). Dindel: Accurate indel calls from short-read data. \emph{Genome Research}, \emph{21}(6), 961--973. doi: \href{https://doi.org/10.1101/gr.112326.110}{10.1101/gr.112326.110}

\leavevmode\DIFaddbegin \hypertarget{ref-Cao_2006}{}%DIF > 
\DIFadd{Cao, Y., Gillespie, D., \& Petzold, L. (2006). Efficient step size selection for the tau-leaping simulation method. }\emph{\DIFadd{The Journal of Chemical Physics}}\DIFadd{, }\emph{124}(4)\DIFadd{, 044109.
}

\leavevmode\hypertarget{ref-Cartwright_2005}{}%DIF > 
\DIFadd{Cartwright, R. A. (2005). DNA assembly with gaps (Dawg): Simulating sequence evolution. }\emph{\DIFadd{Bioinformatics}}\DIFadd{, }\emph{\DIFadd{21}}\DIFadd{, iii31--iii38.
}

\leavevmode\DIFaddend \hypertarget{ref-Chang_2019}{}%
Chang, W. (2019). \emph{R6: Encapsulated classes with reference semantics}. Retrieved from \url{https://CRAN.R-project.org/package=R6}

\leavevmode\DIFaddbegin \hypertarget{ref-Doob_1942}{}%DIF > 
\DIFadd{Doob, J. L. (1942). Topics in the theory of markoff chains. }\emph{\DIFadd{Transactions of the American Mathematical Society}}\DIFadd{, }\emph{52}(1)\DIFadd{, 37--64.
}

\leavevmode\DIFaddend \hypertarget{ref-Eddelbuettel_2011}{}%
Eddelbuettel, D., \& François, R. (2011). Rcpp: Seamless R and C++ integration. \emph{Journal of Statistical Software}, \emph{40}(8), 1--18. doi: \href{https://doi.org/10.18637/jss.v040.i08}{10.18637/jss.v040.i08}

\leavevmode\hypertarget{ref-Escalona_2016}{}%
Escalona, M., Rocha, S., \& Posada, D. (2016). A comparison of tools for the simulation of genomic next-generation sequencing data. \emph{Nature Reviews Genetics}, \emph{17}(8), 459--469. doi: \href{https://doi.org/10.1038/nrg.2016.57}{10.1038/nrg.2016.57}

\leavevmode\hypertarget{ref-Escalona_2018}{}%
Escalona, M., Rocha, S., \& Posada, D. (2018). NGSphy: Phylogenomic simulation of next-generation sequencing data. \emph{Bioinformatics}, \emph{34}(14), 2506--2507. doi: \href{https://doi.org/10.1093/bioinformatics/bty146}{10.1093/bioinformatics/bty146}

\leavevmode\hypertarget{ref-Ewing_2010}{}%
Ewing, G., \& Hermisson, J. (2010). MSMS: A coalescent simulation program including recombination, demographic structure and selection at a single locus. \emph{Bioinformatics}, \emph{26}(16), 2064--2065. doi: \href{https://doi.org/10.1093/bioinformatics/btq322}{10.1093/bioinformatics/btq322}

\leavevmode\hypertarget{ref-Felsenstein_1981}{}%
Felsenstein, J. (1981). Evolutionary trees from DNA sequences: A maximum likelihood approach. \emph{Journal of Molecular Evolution}, \emph{17}(6), 368--376. doi: \href{https://doi.org/10.1007/bf01734359}{10.1007/bf01734359}

\leavevmode\hypertarget{ref-Fletcher_2009}{}%
Fletcher, W., \& Yang, Z. (2009). INDELible: A flexible simulator of biological sequence evolution. \emph{Molecular Biology and Evolution}, \emph{26}(8), 1879--1888. doi: \href{https://doi.org/10.1093/molbev/msp098}{10.1093/molbev/msp098}

\leavevmode\hypertarget{ref-Forner_2018}{}%
Forner, K. (2018). \emph{RcppProgress: An interruptible progress bar with openmp support for c++ in r packages}. Retrieved from \url{https://CRAN.R-project.org/package=RcppProgress}

\leavevmode\DIFaddbegin \hypertarget{ref-Gillespie_1976}{}%DIF > 
\DIFadd{Gillespie, D. T. (1976). A general method for numerically simulating the stochastic time evolution of coupled chemical reactions. }\emph{\DIFadd{Journal of Computational Physics}}\DIFadd{, }\emph{22}(4)\DIFadd{, 403--434.
}

\leavevmode\DIFaddend \hypertarget{ref-Hasegawa_1985}{}%
Hasegawa, M., Kishino, H., \& Yano, T.-a. (1985). Dating of the human-ape splitting by a molecular clock of mitochondrial DNA. \emph{Journal of Molecular Evolution}, \emph{22}(2), 160--174. doi: \href{https://doi.org/10.1007/bf02101694}{10.1007/bf02101694}

\leavevmode\hypertarget{ref-Hasegawa_1984}{}%
Hasegawa, M., Yano, T.-a., \& Kishino, H. (1984). A new molecular clock of mitochondrial DNA and the evolution of hominoids. \emph{Proceedings of the Japan Academy, Series B}, \emph{60}(4), 95--98. doi: \href{https://doi.org/10.2183/pjab.60.95}{10.2183/pjab.60.95}

\leavevmode\hypertarget{ref-Hayden_2019}{}%
Hayden, N., \& Morgan, M. (2019). \emph{Rhtslib: HTSlib high-throughput sequencing library as an \DIFdelbegin \DIFdel{r }\DIFdelend \DIFaddbegin \DIFadd{R }\DIFaddend package}.

\leavevmode\hypertarget{ref-Holtgrewe_2010}{}%
Holtgrewe, M. (2010). Masona read simulator for second generation sequencing data. \emph{Technical Report Freie Universität Berlin}.

\leavevmode\hypertarget{ref-Huang_2011}{}%
Huang, W., Li, L., Myers, J. R., \& Marth, G. T. (2011). ART: A next-generation sequencing read simulator. \emph{Bioinformatics}, \emph{28}(4), 593--594. doi: \href{https://doi.org/10.1093/bioinformatics/btr708}{10.1093/bioinformatics/btr708}

\leavevmode\hypertarget{ref-Hudson_2002}{}%
Hudson, R. R. (2002). Generating samples under a wright-fisher neutral model of genetic variation. \emph{Bioinformatics}, \emph{18}(2), 337--338. doi: \href{https://doi.org/10.1093/bioinformatics/18.2.337}{10.1093/bioinformatics/18.2.337}

\leavevmode\hypertarget{ref-JC69}{}%
Jukes, T. H., \& Cantor, C. R. (1969). Evolution of protein molecules. In H. N. Munro (Ed.), \emph{Mammalian protein metabolism} (Vol. 3, pp. 21--131). New York: Academic Press.

\leavevmode\hypertarget{ref-Kimura_1980}{}%
Kimura, M. (1980). A simple method for estimating evolutionary rates of base substitutions through comparative studies of nucleotide sequences. \emph{Journal of Molecular Evolution}, \emph{16}(2), 111--120. doi: \href{https://doi.org/10.1007/bf01731581}{10.1007/bf01731581}

\leavevmode\DIFdelbegin %DIFDELCMD < \hypertarget{ref-Knaus_2017}{}%%%
%DIF < 
\DIFdel{Knaus, B. J., \& Grünwald, N. J. (2017). VCFR: A package to manipulate and visualize variant call format data in R. }\emph{\DIFdel{Molecular Ecology Resources}}%DIFAUXCMD
\DIFdel{, }%DIFDELCMD < \emph{17}(1)%%%
\DIFdel{, 44--53.
}%DIFDELCMD < 

%DIFDELCMD < \leavevmode%%%
\DIFdelend \hypertarget{ref-Kronmal_1979}{}%
Kronmal, R. A., \& Peterson, A. V. (1979). On the alias method for generating random variables from a discrete distribution. \emph{The American Statistician}, \emph{33}(4), 214--218. doi: \href{https://doi.org/10.1080/00031305.1979.10482697}{10.1080/00031305.1979.10482697}

\leavevmode\hypertarget{ref-Li_2009}{}%
Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., \ldots{} and, R. D. (2009). The sequence alignment/map format and SAMtools. \emph{Bioinformatics}, \emph{25}(16), 2078--2079. doi: \href{https://doi.org/10.1093/bioinformatics/btp352}{10.1093/bioinformatics/btp352}

\leavevmode\hypertarget{ref-Li_2011}{}%
Li, X., Zhu, C., Lin, Z., Wu, Y., Zhang, D., Bai, G., \ldots{} Yu, J. (2011). Chromosome size in diploid eukaryotic species centers on the average length with a conserved boundary. \emph{Molecular Biology and Evolution}, \emph{28}(6), 1901--1911. doi: \href{https://doi.org/10.1093/molbev/msr011}{10.1093/molbev/msr011}

\leavevmode\hypertarget{ref-Mallo_2015}{}%
Mallo, D., Oliveira Martins, L. de, \& Posada, D. (2015). SimPhy: Phylogenomic simulation of gene, locus, and species trees. \emph{Systematic Biology}, \emph{65}(2), 334--344.

\leavevmode\hypertarget{ref-McElroy_2012}{}%
McElroy, K. E., Luciani, F., \& Thomas, T. (2012). GemSIM: General, error-model based simulator of next-generation sequencing data. \emph{BMC Genomics}, \emph{13}(1), 74.

\leavevmode\hypertarget{ref-McTavish_2017}{}%
McTavish, E. J., Pettengill, J., Davis, S., Rand, H., Strain, E., Allard, M., \& Timme, R. E. (2017). TreeToReads - a pipeline for simulating raw reads from phylogenies. \emph{BMC Bioinformatics}, \emph{18}(1). doi: \href{https://doi.org/10.1186/s12859-017-1592-1}{10.1186/s12859-017-1592-1}

\leavevmode\hypertarget{ref-Metzker_2009}{}%
Metzker, M. L. (2009). Sequencing technologies the next generation. \emph{Nature Reviews Genetics}, \emph{11}(1), 31--46. doi: \href{https://doi.org/10.1038/nrg2626}{10.1038/nrg2626}

\leavevmode\hypertarget{ref-Morgan_2019}{}%
Morgan, M. (2019). \emph{Zlibbioc: An r packaged zlib-1.2.5}. Retrieved from \url{http://bioconductor.org/packages/release/bioc/html/Zlibbioc.html}

\leavevmode\hypertarget{ref-Oneill_2014pcg}{}%
O'Neill, M. E. (2014). \emph{PCG: a family of simple fast space-efficient statistically good algorithms for random number generation}. Claremont, CA: Harvey Mudd College.

\leavevmode\hypertarget{ref-Paradis_2018}{}%
Paradis, E., \& Schliep, K. (2018). Ape 5.0: An environment for modern phylogenetics and evolutionary analyses in R. \emph{Bioinformatics}, \emph{35}, 526--528.

\leavevmode\hypertarget{ref-Rambaut_1997}{}%
Rambaut, A., \& \DIFdelbegin \DIFdel{Grass}\DIFdelend \DIFaddbegin \DIFadd{Grassly}\DIFaddend , N. C. (1997). Seq-gen: An application for the monte carlo simulation of DNA sequence evolution along phylogenetic trees. \emph{\DIFdelbegin \DIFdel{Bioinformatics}\DIFdelend \DIFaddbegin \DIFadd{Computer Applications in the Biosciences}\DIFaddend }, \emph{13}(3), 235--238. doi: \href{https://doi.org/10.1093/bioinformatics/13.3.235}{10.1093/bioinformatics/13.3.235}

\leavevmode\hypertarget{ref-R_Core_Team_2019}{}%
R Core Team. (2019). \emph{R: a language and environment for statistical computing}. Retrieved from \url{https://www.R-project.org/}

\leavevmode\hypertarget{ref-Paul_R._Staab_2016}{}%
Staab, P. R., \& Metzler, D. (2016). Coala: An R framework for coalescent simulation. \emph{Bioinformatics}. doi: \href{https://doi.org/10.1093/bioinformatics/btw098}{10.1093/bioinformatics/btw098}

\leavevmode\hypertarget{ref-Paul_R._Staab_2015}{}%
Staab, P. R., Zhu, S., Metzler, D., \& Lunter, G. (2015). scrm: Efficiently simulating long sequences using the approximated coalescent with recombination. \emph{Bioinformatics}, \emph{31}(10), 1680--1682.

\leavevmode\hypertarget{ref-St_cker_2016}{}%
Stöcker, B. K., Köster, J., \& Rahmann, S. (2016). SimLoRD: Simulation of long read data. \emph{Bioinformatics}, \emph{32}(17), 2704--2706. doi: \href{https://doi.org/10.1093/bioinformatics/btw286}{10.1093/bioinformatics/btw286}

\leavevmode\DIFdelbegin %DIFDELCMD < \hypertarget{ref-Sung_2016}{}%%%
%DIF < 
\DIFdel{Sung, W., Ackerman, M. S., Dillon, M. M., Platt, T. G., Fuqua, C., Cooper, V. S., \& Lynch, M. (2016). Evolution of the insertion-deletion mutation rate across the tree of life. }\emph{\DIFdel{G3: Genes\(\vert\)Genomes\(\vert\)Genetics}}%DIFAUXCMD
\DIFdel{, }%DIFDELCMD < \emph{6}(8)%%%
\DIFdel{, 2583--2591. doi: }%DIFDELCMD < \href{https://doi.org/10.1534/g3.116.030890}{10.1534/g3.116.030890}
%DIFDELCMD < 

%DIFDELCMD < \leavevmode%%%
\DIFdelend \hypertarget{ref-TN93}{}%
Tamura, K., \& Nei, M. (1993). Estimation of the number of nucleotide substitutions in the control region of mitochondrial dna in humans and chimpanzees. \emph{Molecular Biology and Evolution}, \emph{10}(3), 512--526.

\leavevmode\hypertarget{ref-Tavare_1986gtr}{}%
Tavar\a'e, S. (1986). Some probabilistic and statistical problems in the analysis of DNA sequences. \emph{Lectures on Mathematics in the Life Sciences}, \emph{17}(2), 57--86.

\leavevmode\hypertarget{ref-Thorne_1992}{}%
Thorne, J. L., Kishino, H., \& Felsenstein, J. (1992). Inching toward reality: an improved likelihood model of sequence evolution. \emph{Journal of Molecular Evolution}, \emph{34}(1), 3--16.

\leavevmode\hypertarget{ref-Thurmond_2018}{}%
Thurmond, J., Goodman, J. L., Strelets, V. B., Attrill, H., Gramates, L. S., Marygold, S. J., \ldots{} \DIFdelbegin \DIFdel{and}\DIFdelend \DIFaddbegin \DIFadd{Baker}\DIFaddend , P. \DIFdelbegin \DIFdel{B. }\DIFdelend (2018). FlyBase 2.0: The next generation. \emph{Nucleic Acids Research}, \emph{47}(D1), D759--D765. doi: \href{https://doi.org/10.1093/nar/gky1003}{10.1093/nar/gky1003}

\leavevmode\hypertarget{ref-Walker_1974}{}%
Walker, A. (1974). New fast method for generating discrete random numbers with arbitrary frequency distributions. \emph{Electronics Letters}, \emph{10}(8), 127. doi: \href{https://doi.org/10.1049/el:19740097}{10.1049/el:19740097}

\leavevmode\DIFaddbegin \hypertarget{ref-Wieder_2011}{}%DIF > 
\DIFadd{Wieder, N., Fink, R., \& Wegner, F. von. (2011). Exact and approximate stochastic simulation of intracellular calcium dynamics. }\emph{\DIFadd{Journal of Biomedicine and Biotechnology}}\DIFadd{, }\emph{\DIFadd{2011}}\DIFadd{, 572492.
}

\leavevmode\DIFaddend \hypertarget{ref-Yang_2006}{}%
Yang, Z. (2006). \emph{Computational molecular evolution}. doi: \href{https://doi.org/10.1093/acprof:oso/9780198567028.001.0001}{10.1093/acprof:oso/9780198567028.001.0001}

\leavevmode\hypertarget{ref-Yang_1994}{}%
Yang, Z. B. (1994). Estimating the pattern of nucleotide substitution. \emph{Journal of Molecular Evolution}, \emph{39}(1), 105--111.

\setstretch{2}

\hypertarget{data-accessibility}{%
\section{Data Accessibility}\label{data-accessibility}}

\texttt{jackalope} is open source, under the MIT license.
\DIFdelbegin \DIFdel{Its code }\DIFdelend \DIFaddbegin \DIFadd{The stable version of }\texttt{\DIFadd{jackalope}} \DIFaddend is available on \DIFdelbegin \DIFdel{GitHub
(}%DIFDELCMD < \url{https://github.com/lucasnell/jackalope}%%%
\DIFdelend \DIFaddbegin \DIFadd{CRAN
(}\url{https://CRAN.R-project.org/package=jackalope}\DIFaddend ),
and the \DIFaddbegin \DIFadd{development version is on GitHub
(}\url{https://github.com/lucasnell/jackalope}\DIFadd{).
The }\DIFaddend documentation can be found at \url{https://jackalope.lucasnell.com}.
The \DIFdelbegin \DIFdel{package has been submitted to CRAN.
The }\DIFdelend version used in this manuscript was \DIFdelbegin \DIFdel{0.1}\DIFdelend \DIFaddbegin \DIFadd{1.0}\DIFaddend .0.
Code for the example usage, validation, and performance is available on GitHub at
\url{https://github.com/lucasnell/jlp_ms}.

\hypertarget{author-contributions}{%
\section{Author Contributions}\label{author-contributions}}

L.A.N. conceived and designed the project, wrote the code, and wrote the manuscript.

\begin{figure}
\centering
\includegraphics{Fig_1.pdf}
\caption{\label{fig:jackalope-overview-figure}Overview of primary \texttt{jackalope} functions, classes, inputs, and outputs. Circles \texttt{sub\_models} and \texttt{vars\_functions} refer to multiple functions; see text for details. \(\theta\) indicates the population-scaled mutation rate.}
\end{figure}

\begin{figure}
\centering
\includegraphics{_manuscript_files/figure-latex/variants-perf-test-plot-1.pdf}
\caption{\label{fig:variants-perf-test-plot}Performance comparison between \texttt{jackalope} and \DIFdelbeginFL \texttt{\DIFdelFL{Seq-Gen}} %DIFAUXCMD
\DIFdelendFL \DIFaddbeginFL \texttt{\DIFaddFL{INDELible}} \DIFaddendFL in generating variants from a reference genome. Sub-panel columns indicate the size of the genome from which variants were generated. \DIFaddbeginFL \DIFaddFL{Sub-panel rows indicate the maximum tree depth of the phylogeny used for the simulations. }\DIFaddendFL Superscripts in the y-axis labels indicate the number of threads used.}
\end{figure}

\begin{figure}
\centering
\includegraphics{_manuscript_files/figure-latex/sequencing-perf-test-plot-1.pdf}
\caption{\label{fig:sequencing-perf-test-plot}Performance comparison between \texttt{jackalope} and (A) \texttt{ART} in generating Illumina reads and (B) \texttt{SimLoRD} in generating PacBio reads. Sub-panel columns indicate the number of reads generated. Superscripts in the y-axis labels indicate the number of threads used.}
\DIFaddbeginFL \end{figure}

\begin{figure}
\centering
\includegraphics{_manuscript_files/figure-latex/pipeline-perf-test-plot-1.pdf}
\caption{\label{fig:pipeline-perf-test-plot}\DIFaddFL{(A) Speed and (B) memory usage comparison between }\texttt{\DIFaddFL{jackalope}} \DIFaddFL{and }\texttt{\DIFaddFL{NGSphy}} \DIFaddFL{in generating variants from a reference genome, then producing Illumina reads from those variants. In (A), sub-panel columns indicate the maximum tree depth of the phylogeny used for the simulations. In both plots, sub-panel rows indicate the size of the genome from which variants were generated, and color indicates the program used; superscripts in the color labels indicate the number of threads used. The \({}^{*}\) highlights where results were combined for calls to }\texttt{\DIFaddFL{NGSphy}} \DIFaddFL{with 1 and 4 threads; see text for details.}}
\DIFaddendFL \end{figure}




\end{document}
