---
output: html_document
editor_options: 
  chunk_output_type: console
---


This tests that the PacBio sequencer produces quality profiles and read lengths similar
to expectations.



```{r pacbio-setup, include=FALSE, cache = FALSE, eval = FALSE}
suppressPackageStartupMessages({
    library(jackalope)
    library(tidyverse)
})
knitr::opts_chunk$set(echo = FALSE, cache = FALSE, dev = 'quartz_pdf')
# knitr::opts_knit$set(root.dir = normalizePath(".."))
if (isTRUE(getOption('knitr.in.progress'))) knitr::knit_theme$set(knitr::knit_theme$get("acid"))
if (!isTRUE(getOption('knitr.in.progress'))) {
    options("device" = "quartz")
    grDevices::graphics.off()
}
theme_set(theme_classic() + 
              theme(legend.background = element_blank(),
                    strip.background = element_blank(),
                    plot.margin = margin(0.5, 0.5, 0.5, 0.5)))

# Saves our temporary directory to store some files:
dir <- paste0(tempdir(check = TRUE), "/")

```




## Read lengths - default parameter values

For this section, I simulated 100,000 PacBio reads under default parameters.
I then compared the read lengths to the expected read-length distribution:

$$
\log(L - \delta) \sim \text{N}(\mu, \sigma)
$$

for read lengths $L \ge 50$, where $\delta = 10075.44$, $\mu = 9.79$,
and $\sigma = 0.20$.


```{r pacbio-read-lengths-cap}
set.seed(79164)
rl_cap <- paste("Read length distribution observed from 100,000 generated reads",
                "(histogram; left y-axis), alongside the frequency distribution",
                "as predicted for default parameters (dashed line; right y-axis).")
```


```{r pacbio-read-lengths, fig.height=3, fig.width=5, fig.cap=rl_cap}
# 4 sequences of length 100e3
seq <- sapply(c("T", "C", "A", "G"), function(nt) paste(rep(nt, 100e3), collapse = ""))

# Make ref_genome object from a pointer to a RefGenome object based on `seqs`
rg <- ref_genome$new(jackalope:::make_ref_genome(seq))

pacbio(rg, out_prefix = paste0(dir, "test_pb"), n_reads = 10e3)

fq <- readLines(paste0(dir, "test_pb_R1.fq"))
read_lens <- nchar(fq[seq(2, length(fq), 4)])


# Looking at default read length distribution:

# Defaults from SimLoRD:
test_lens <- function(obs, mult = 1) {
    pars <- list(sigma = 0.200110276521, loc = -10075.4363813, scale = 17922.611306,
                 min_len = 50)
    .x <- seq(pars$min_len, max(obs), length.out = length(obs))
    .z <- mult * dlnorm(.x - pars$loc, log(pars$scale), pars$sigma)
    return(.z)
}


tibble(obs = nchar(fq[seq(2, length(fq), 4)])) %>%
    ggplot(aes(obs)) +
    # geom_freqpoly(bins = 50, size = 1, color = "dodgerblue") +
    geom_histogram(bins = 50, fill = "dodgerblue", color = "gray70", size = 0.1) +
    stat_function(fun = test_lens, args = list(mult = 6.5e6), color = "black",
                  linetype = 2, size = 1) +
    theme(legend.position = c(0.75, 0.75),
          legend.text = element_text(size = 11)) +
    scale_color_brewer(NULL, palette = "Dark2") +
    # ggtitle("Read length distributions") +
    scale_y_continuous("Count",
                       sec.axis = sec_axis(~./6.5e6, name = "Density")) +
    scale_x_continuous("Read length (kb)",
                       breaks = seq(0, 30e3, 10e3), labels = seq(0, 30, 10))
```




## Read lengths - custom distribution

In this case, we generated 100,000 PacBio reads with a custom read-length distribution.
The possible read lengths for this distribution were in the sequence from 100 to 10,000
by increments of 100.
Each read length had a sampling weight drawn from a uniform distribution.
So for a vector of weights ($W$) corresponding to each length, the number of predicted
reads of the $i^{\text{th}}$ length is simply

$$
N \frac{ W_i }{ \sum{W} }
\text{,}
$$
where $N$ is the number of reads.


```{r pacbio-read-lengths-non-defaults-cap}
rl_nd_cap <- paste("Observed versus predicted number of reads for PacBio reads",
                   "generated from a custom distribution.",
                   "The dashed line is the 1-to-1 line.")
```


```{r pacbio-read-lengths-non-defaults, fig.cap=rl_nd_cap}

# Looking at non-default read length distribution

rl_mat <- cbind(seq(100, 1e4, 100), runif(100))
rl_mat[,2] <- rl_mat[,2] / sum(rl_mat[,2])
pacbio(rg, out_prefix = paste0(dir, "test_pb_nd"), n_reads = 10e3,
       custom_read_lengths = rl_mat, overwrite = TRUE)

fq_nd <- readLines(paste0(dir, "test_pb_nd_R1.fq"))
read_lens <- nchar(fq_nd[seq(2, length(fq_nd), 4)])


# Plotting observed vs predicted:

ndrl_df <- tibble(obs = as.integer(table(factor(read_lens, levels = rl_mat[,1]))),
                  pred = length(read_lens) * rl_mat[,2])

ndrl_df %>%
    ggplot(aes((pred), (obs))) +
    geom_point() +
    geom_abline(slope = 1, intercept = 0, linetype = 2, color = "firebrick2") +
    scale_x_continuous("Predicted read length count") +
    scale_y_continuous("Observed read length count") +
    coord_equal()

```





## Read qualities

Here I compared average read quality and read length between `jackalope` and `SimLoRD`
by comparing the previous simulations performed in `jackalope` with
output from the following `SimLoRD` command:

```
simlord --read-reference ref.fa -n 10000 --no-sam simlord_out
```

where `ref.fa` is a FASTA file written from the reference used in the `jackalope`
simulations.


```{r pacbio-write-reference, eval = FALSE}
# I did this first for the SimLoRD simulations:
write_fasta(rg, "perf_tests/diag_ref")
```
```{bash pacbio-run-simlord, eval = FALSE}
## How I ran the SimLoRD simulations:
conda activate base
source activate simlord

cd ~/GitHub/Wisconsin/jlp_ms/perf_tests

simlord --read-reference diag_ref.fa -n 10000 --no-sam diag_pb_out
```


```{r pacbio-pb-q-cap}
pb_q_cap <- paste("Read length and average quality per read for 10,000",
                  "PacBio reads simulated using \\texttt{jackalope} and",
                  "\\texttt{SimLoRD}.",
                  "Color of hex bins indicates the number of occurrences of that",
                  "combination of read length and average quality.")
```


```{r pacbio-quals, fig.width=7.5, fig.height=4, fig.cap=pb_q_cap}
Rcpp::sourceCpp("supp/diagnostics-pacbio.cpp")

# --------*
# Jackalope:
# --------*
fq <- readLines(paste0(dir, "test_pb_R1.fq"))
read_lens <- nchar(fq[seq(2, length(fq), 4)])

quals <- fq[seq(4, length(fq), 4)]
# Mean quality by read:
m_quals <- mean_quals(quals)
pb_jlp <- tibble(qual = m_quals, len = read_lens) %>% 
    mutate(method = "jackalope")


# --------*
# SimLoRD:
# --------*
fq <- readLines("perf_tests/diag_pb_out.fastq")
read_lens <- nchar(fq[seq(2, length(fq), 4)])

quals <- fq[seq(4, length(fq), 4)]
# Mean quality by read:
m_quals <- mean_quals(quals)
pb_sl <- tibble(qual = m_quals, len = read_lens) %>% 
    mutate(method = "SimLoRD")


bind_rows(pb_jlp, pb_sl) %>% 
    mutate(method = factor(method)) %>%
    ggplot(aes(len / 1000, qual)) +
    geom_hex(bins = 50) +
    facet_wrap(~ method, ncol = 2) +
    xlab("Read length (kb)") +
    ylab("Read quality\n(average base quality)") +
    theme(legend.position = c(0.85, 0.7), strip.background = element_blank())

```