---
editor_options: 
  chunk_output_type: console
---


```{r illumina-use_rds, echo = FALSE, eval = TRUE}
# Logical for whether to use saved RDS files. 
# Alternative is to re-run everything.
use_rds <- TRUE
```

```{r illumina-setup, include=FALSE, cache = FALSE, eval = FALSE}
suppressPackageStartupMessages({
    library(knitr)
    library(knitcitations)
    library(bookdown)
    library(jackalope)
    library(tidyverse)
})
knitr::opts_chunk$set(echo = FALSE, cache = FALSE, dev = 'quartz_pdf')
# knitr::opts_knit$set(root.dir = normalizePath(".."))
if (isTRUE(getOption('knitr.in.progress'))) {
  knitr::knit_theme$set(knitr::knit_theme$get("acid"))
} 
cleanbib()
options("citation_format" = "pandoc")
if (!isTRUE(getOption('knitr.in.progress'))) {
    options("device" = "quartz")
    grDevices::graphics.off()
}
theme_set(theme_classic() + 
              theme(legend.background = element_blank(),
                    strip.background = element_blank(),
                    plot.margin = margin(0.5, 0.5, 0.5, 0.5)))

# Saves our temporary directory to store some files:
dir <- paste0(tempdir(check = TRUE), "/")
```


This section tests that the Illumina sequencer produces output that fits our
expectations.










## Testing mismatches and quality scores:


```{r illumina-prof-pred-sourceCpp}
set.seed(94861548)
Rcpp::sourceCpp("supp/diagnostics-illumina.cpp")

# Profile I'll be using
prof <- jackalope:::read_profile(NULL, "HS25", read_length = 100, 1)

# Predicted qualities for each nucleotide
pred_df <- map_dfr(1:4,
                   function(nti) {
                       map_dfr(1:100,
                               function(i) {
                                   cbind(nti, i, prof$quals[[nti]][[i]],
                                         prof$qual_probs[[nti]][[i]]) %>%
                                       as.data.frame() %>%
                                       as_tibble() %>%
                                       set_names(c("nucleo", "pos", "qual", "prob")) %>%
                                       mutate_at(vars(pos, qual), as.integer) %>%
                                       mutate(nucleo = c("T", "C", "A", "G")[nti])
                               })
                   })

bases <- c("T", "C", "A", "G")

# 1 sequence of length 100e3
seq <- sapply(bases, function(nt) paste(rep(nt, 100e3), collapse = ""))

# Make ref_genome object from a pointer to a RefGenome object based on `seqs`
rg <- ref_genome$new(jackalope:::make_ref_genome(seq))


```


Mismatches and quality scores are simulated the same as in ART.
First, qualities profiles indicate the frequency of quality scores by position on read.
Then, the probability of a mismatch for quality $q$ is $10^{- q / 10}$.
I am testing the accuracy of the output from `illumina` by starting with a
reference genome consisting of 4 sequences, each containing a single nucleotide
repeated 100,000 times:

```{r illumina-print-ref}
rg
```



Then I simulated 100,000 100-bp reads from the HiSeq 2500 platform ten times.
Reads were matched with the sequence they derived from based on which nucleotide
comprised the majority of the read.
I counted mismatches as nucleotides that differ from the majority nucleotide.
Qualities are directly measured from the output FASTQ.
Both mismatches and qualities are aggregated by their position on the read,
and compared to expectations as calculated from the profile file.


```{r illumina-test-mm-quals}

set.seed(78915)
for (i in 1:10) {
    illumina(rg, out_prefix = paste0(dir, "test", i),
             n_reads = 100e3, read_length = 100, 
             paired = FALSE, seq_sys = "HS25",
             frag_mean = 400, frag_sd = 100, 
             ins_prob1 = 0, del_prob1 = 0, overwrite = TRUE)
}; rm(i)

fastqs <- lapply(1:10, function(i) readLines(paste0(dir, "test", i, "_R1.fq")))
reads <- lapply(fastqs, function(fq) fq[seq(2, length(fq), 4)])
quals <- lapply(fastqs, function(fq) fq[seq(4, length(fq), 4)])
rm(fastqs) # uses a bunch of memory
nts <- lapply(reads, function(rds) which_nt(rds) + 1)

ill_plot1_cap <- paste("The proportion of reads derived from each reference genome",
                       "sequence. Since all sequences were the same size, reads",
                       "were expected to be generated from each sequence with equal",
                       "probability (indicated by the horizontal gray dashed line).")
```


```{r illumina-test-mm-qual-plots1, fig.height=3, fig.width=5, fig.cap=ill_plot1_cap}
# Should be ~0.25 each:
map_dfr(1:10, function(i) tibble(x = 1:4, y = map_dbl(1:4, ~ mean(nts[[i]] == .x)))) %>% 
    mutate(x = factor(x, levels = 1:4, labels = rg$chrom_names())) %>% 
    ggplot() + 
    geom_hline(yintercept = 0.25, linetype = 2, color = "gray70") + 
    geom_jitter(aes(x, y, color = x), height = 0, width = 0.2) +
    scale_color_viridis_d(option = "D", guide = FALSE) +
    theme(axis.text.x = element_text(color = "black", size = 10),
          axis.title.x = element_blank()) +
    ylab("Proportion of reads")
```


```{r illumina-test-mm-qual-plots2-cap}
ill_plot2_cap <- paste("Observed versus predicted quality scores for each nucleotide.",
                       "The dashed lines are the 1-to-1 lines.",
                       "Each point represents the number of times a quality score",
                       "was present at a given position and nucleotide for each",
                       "of the 10 simulations.")
```


```{r illumina-test-mm-qual-plots2, fig.cap=ill_plot2_cap, fig.height=5, fig.width=5.5}
#'
#' Looking at mapping qualities as predicted by position on read and
#' starting nucleotide
#'

if (use_rds) {
    qbp_df <- readRDS("supp/rds_files/qbp_df.rds")
} else {
    # Takes ~4 min
    qbp_df <-
        pmap_dfr(
            crossing(nt = 1:4, i = 1:10),
            function(nt, i) {
                do.call(cbind, quals_by_pos(quals[[i]][nts[[i]] == nt])) %>%
                    as.data.frame() %>%
                    as_tibble() %>%
                    set_names(sprintf("p%i", 1:100)) %>%
                    gather("position", "quality") %>%
                    mutate(position = as.integer(gsub("p", "", position))) %>%
                    group_by(position, quality) %>%
                    summarize(count = n(),
                              pred_count = sum(nts[[i]] == nt) *
                                  pred_df$prob[pred_df$pos == position[1] &
                                                   pred_df$qual == quality[1] &
                                                   pred_df$nucleo == bases[nt]]) %>%
                    ungroup() %>%
                    mutate(nucleo = bases[nt])
            }) %>%
        mutate(nucleo = factor(nucleo, levels = bases))
    saveRDS(qbp_df, "supp/rds_files/qbp_df.rds")
}


qbp_df %>%
    ggplot(aes(log(pred_count), log(count))) +
    geom_point(alpha = 0.25, shape = 1) +
    geom_abline(slope = 1, intercept = 0, linetype = 2, color = "firebrick2") +
    theme(strip.background = element_blank(),
          strip.text = element_text(face = "bold", size = 12)) +
    facet_wrap(~ nucleo, nrow = 2) +
    scale_y_continuous("Observed quality count", breaks = log(40^(0:3)),
                       labels = 40^(0:3),
                       limits = log(range(c(qbp_df$count, qbp_df$pred_count)))) +
    scale_x_continuous("Predicted quality count", breaks = log(40^(0:3)),
                       labels = 40^(0:3),
                       limits = log(range(c(qbp_df$count, qbp_df$pred_count)))) +
  coord_equal()
```






```{r illumina-test-mm-qual-plots3-cap}
ill_plot3_cap <- paste("Observed versus predicted mismatches.",
                       "The dashed line is the 1-to-1 line.",
                       "Each point represents the number of mismatches",
                       "for a given quality, for each of the 10 simulations.")
```

```{r illumina-test-mm-qual-plots3, fig.cap=ill_plot3_cap}
#'
#' Now looking at mismatches as predicted by quality:
#'

qbp_all <- lapply(1:10, function(i) do.call(c, quals_by_pos(quals[[i]])))
qbp_all <- do.call(cbind, qbp_all)

# Takes ~48 sec
mmbq_df <- map_dfr(1:10,
                   function(i) {
                       tibble(qual = sort(unique(qbp_all[,i])),
                              mm = mm_by_qual(reads[[i]], quals[[i]], 
                                              max_qual = 50)[qual+1],
                              rep = i) %>%
                           mutate(n = map_dbl(qual, ~ sum(qbp_all[,i] == .x)),
                                  p = 10^(qual / -10.0),
                                  mm_pred = n * p)
                    })

mmbq_df2 <- mmbq_df %>% 
  filter(mm > 0) # this only filters one row and doesn't change the conclusions

mmbq_df2 %>% 
  ggplot(aes(log(mm_pred), log(mm))) +
  geom_point(alpha = 0.5, shape = 1) +
  geom_abline(slope = 1, intercept = 0, linetype = 2, color = "firebrick2") +
  scale_x_continuous("Predicted mismatch count", breaks = log(10^(1:3)),
                     labels = 10^(1:3), 
                     limits = log(range(c(mmbq_df2$mm, mmbq_df2$mm_pred)))) +
  scale_y_continuous("Observed mismatch count", breaks = log(10^(1:3)),
                     labels = 10^(1:3), 
                     limits = log(range(c(mmbq_df2$mm, mmbq_df2$mm_pred)))) +
  coord_equal()
```








## Testing indel rates


Indels are simulated the same as in ART, where there is a standard rate at which
indels occur that does not depend on the position on the read.
I am testing the accuracy of the output from `illumina` by starting with a
reference genome with many small (smaller than the read length), same-sized sequences:

```{r illumina-reference}
# 100e3 sequences of length 10
seqs <- replicate(100e3, paste(rep("TA", 10 / 2), collapse = ""))

# Make ref_genome object from a pointer to a RefGenome object based on `seqs`
rg <- ref_genome$new(jackalope:::make_ref_genome(seqs))

rg
```



Then I simulated 100-bp reads using a fake profile with no mismatches.
Indels were found when the read length differed from the sequence length,
and because I tested insertions and deletions separately, there was no chance of
multiple indels offsetting one another.
The predicted number of reads with at least one indel was simply the product of 
the indel rate, read length, and number of reads.



```{r illumina-fake-ill-prof}
# Making fake profile for no mismatches:
profile_df <- crossing(nucleo = c("T", "C", "A", "G"),
                       pos = 0:99,
                       qual = c(255L, 1000L)) %>%
    arrange(nucleo, pos, qual)

write_tsv(profile_df, path = paste0(dir, "test_prof.txt"), col_names = FALSE)
```



```{r illumina-indel-rates}

indel_test <- function(.ins_prob, .del_prob) {
    stopifnot(.ins_prob == 0 | .del_prob == 0)
    stopifnot(.ins_prob > 0 | .del_prob > 0)
    illumina(rg, out_prefix = paste0(dir, "test"),
             n_reads = 100e3, read_length = 100, paired = FALSE,
             profile1 = paste0(dir, "test_prof.txt"),
             frag_mean = 400, frag_sd = 100,
             ins_prob1 = .ins_prob,
             del_prob1 = .del_prob,
             overwrite = TRUE)
    fq <- readLines(paste0(dir, "test_R1.fq"))
    .lens <- nchar(fq[seq(2, length(fq), 4)])
    if (.ins_prob > 0) {
        .type = "insertion"
        .n <- sum(.lens - 10)
        .n_pred <- .ins_prob * 10 * 100e3
    } else if (.del_prob > 0) {
        .type = "deletion"
        .n <- sum(10 - .lens)
        .n_pred <- .del_prob * 10 * 100e3
    }
    return(tibble(type = .type, n = .n, n_pred = .n_pred))
}

indel_rates_plot_cap <- paste("Observed versus predicted number of reads with indels.",
                              "The dashed line is the 1-to-1 line.",
                              "Note that all points overlap.")
```


```{r illumina-indel-rates-plot, fig.cap=indel_rates_plot_cap}

if (use_rds) {
    indel_df <- readRDS("supp/rds_files/indel_df.rds")
} else {
    # Takes ~ 6 min
    indel_df <- crossing(ins_prob = c(0, 10^(seq(-1,-4,length.out = 11))),
                         del_prob = c(0, 10^(seq(-1,-4,length.out = 11)))) %>%
        filter((ins_prob == 0 | del_prob == 0) & (ins_prob > 0 | del_prob > 0)) %>%
        pmap_dfr(~ indel_test(..1, ..2))
    saveRDS(indel_df, "supp/rds_files/indel_df.rds")
}



lims <- log(range(c(indel_df$n, indel_df$n_pred)))
labs <- 10^(2:5)

indel_df %>%
    ggplot(aes(log(n_pred), log(n))) +
    geom_abline(slope = 1, intercept = 0, linetype = 2, color = "firebrick") +
    geom_point(aes(color = type), alpha = 0.5, size = 3) +
    scale_x_continuous("Predicted indel count", limits = lims,
                       breaks = log(labs), labels = labs) +
    scale_y_continuous("Observed indel count", limits = lims,
                       breaks = log(labs), labels = labs) +
    scale_color_brewer(NULL, palette = "Dark2") +
    theme(legend.position = c(0.3, 0.9)) +
    coord_equal()
```

