---
editor_options: 
  chunk_output_type: console
---


```{r mevo-setup, include=FALSE, cache = FALSE, eval = FALSE}
suppressPackageStartupMessages({
    library(jackalope)
    library(tidyverse)
    library(grid)
    library(ape)
})
knitr::opts_chunk$set(echo = FALSE, cache = FALSE, dev = 'quartz_pdf')
# knitr::opts_knit$set(root.dir = normalizePath(".."))
if (isTRUE(getOption('knitr.in.progress'))) {
  knitr::knit_theme$set(knitr::knit_theme$get("acid"))
}
if (!isTRUE(getOption('knitr.in.progress'))) {
    options("device" = "quartz")
    grDevices::graphics.off()
}
theme_set(theme_classic() + 
              theme(legend.background = element_blank(),
                    strip.background = element_blank(),
                    plot.margin = margin(0.5, 0.5, 0.5, 0.5)))

# Saves our temporary directory to store some files:
dir <- paste0(tempdir(check = TRUE), "/")
```


This tests that when creating variants, `jackalope` produces substitutions, indels,
and among-site variation that fit expectations.
It lastly checks that when simulating along a phylogeny, the number of 
differences between individual variants is proportional to the phylogenetic distance
between them.

To conduct these tests, `jackalope` is compiled with `__JACKALOPE_DIAGNOSTICS`
defined inside the `src/jackalope_config.h` file.
This causes `jackalope` to print information about mutations as they are produced and
about among-site variation as the rates are generated.
In most of the sections below, I captured this console output to describe the 
observed output from `jackalope`.

The reference genome used in this section is displayed below:

```{r mevo-simulate}

set.seed(1087437799)

# simulate ----

# Set all needed molecular evolution parameters inside an environment (using TN93 method)
pars <- new.env()
with(pars, {
    n_chroms <- 20
    chrom_len <- 100e3
    # Molecular evolution parameters:
    pi_tcag <- c(0.1, 0.2, 0.3, 0.4)
    alpha_1 <- 0.25
    alpha_2 <- 0.35
    beta <- 0.5
    #   For indels:
    rates <- c(0.2, 0.3)
    M <- c(8L, 10L)
    a <- c(0.1, 0.5)
    n_vars <- 10
    #   For site variation:
    shape <- 0.5
    # Construct chromosomes with known number of each nucleotide
    chroms <- rep(list(c(rep("T", 0.25 * chrom_len), rep("C", 0.25 * chrom_len),
                       rep("A", 0.25 * chrom_len), rep("G", 0.25 * chrom_len))),
                n_chroms)
    chroms <- lapply(chroms, sample)
    chroms <- sapply(chroms, paste, collapse = "")
})

# Make ref_genome object from a pointer to a RefGenome object based on `chroms`
ref <- with(pars, ref_genome$new(jackalope:::make_ref_genome(chroms)))
ref
```





## Substitution rates

```{r illumina-sub-cap-inputs}
# Phylogenetic tree:
tree <- ape::rcoal(2)
tree$edge.length <- rep(0.1, 2)

sub <- sub_TN93(alpha_1 = pars$alpha_1,
                alpha_2 = pars$alpha_2,
                beta = pars$beta,
                pi_tcag = pars$pi_tcag)

ill_sub_cap <- paste("Observed versus predicted substitutions for simulation of",
                     "2 variants along a phylogeny.")
```

This simulates just substitutions along a phylogeny of 2 variants, with
a branch length of `r signif(tree$edge.length[1], 3)` and with the following
substitution rate matrix:

```{r mevo-sub-mat}
prmatrix(sub$Q()[[1]], digits = 3,
         rowlab = paste("  ", c("T", "C", "A", "G")),
         collab = c("T", "C", "A", "G"))
```


I examined the substitutions from the output object, rather than from console diagnostics,
because it was simpler this way since there was no among-site variation.
I counted the number of each type of mutation (corresponding to all non-diagonal 
elements in the rate matrix) for each chromosome and variant.
I compared these to the predicted substitution counts, which were the product of the
chromosome size and the transition-probability matrix for the branch length of
the phylogeny.


```{r mevo-sub-rates, fig.cap=ill_sub_cap, results="hide"}

# Only substitutions
var_set <- create_variants(ref, vars_phylo(tree), sub = sub)

# ---------*
# Observed values:
# ---------*
sub_mats <- map(0:(length(tree$tip.label)-1),
                function(v) {
                  map(0:(pars$n_chroms-1),
                       function(s) {
                           Z <- jackalope:::examine_mutations(var_set_ptr = var_set$ptr(),
                                                       var_ind = v, chrom_ind = s) %>%
                           .[["sub"]]
                       return(Z)
                   })
                }) %>% 
  Reduce(f = c)

# ---------*
# Predicted values:
# ---------*
# Calculate probability transition matrix
Pt <- function(t, .sub, i = 1) {
  diag_L <- diag(as.numeric(exp(.sub$L()[[i]] * t)))
  .sub$U()[[i]] %*% diag_L %*% .sub$Ui()[[i]]
}
# Expected # mutations per variant per chromosome, of each type
e_muts <- Pt(max(ape::node.depth.edgelength(tree)), sub) * # mutations per base
  matrix(ref$sizes()[1], 4, 4) *  # going from mutations/base to mutation/chromosome
  matrix(0.25, 4, 4) *  # accounting for nucleotide frequencies in ref. genome
  (1 - diag(4))  # setting diagonal to zero


process_sub_mat <- function(sm, .id) {
  .df <- tibble(
    n = c(sm[lower.tri(sm)], sm[upper.tri(sm)]),
    to = c(rep(c("T","C","A","G"), 3:0), 
            rep(c("T","C","A","G"), 0:3)),
    from = c(c("T","C","A","G")[c(2:4, 3:4, 4)],
              c("T","C","A","G")[c(1, 1:2, 1:3)]),
    id = .id) %>% 
    mutate_at(vars(to, from), factor, levels = c("T","C","A","G"))
  return(.df)
}

sub_df <- map2_dfr(sub_mats, 1:length(sub_mats), process_sub_mat) %>% 
  mutate(pred = rep(process_sub_mat(e_muts, "exp")[["n"]], length(sub_mats))) %>% 
  rename(obs = n)


sub_df %>%
    ggplot(aes(pred, obs)) +
    geom_point(shape = 1) +
    geom_abline(slope = 1, intercept = 0, linetype = 2, color = "firebrick2") +
    scale_x_continuous("Predicted substitution count") +
    scale_y_continuous("Observed substitution count") +
    coord_equal()
```



## Indel rates

In this section, I first set the deletion rate to zero and
the insertion rate to `r pars$rates[1]`.
In the second set of simulations, the insertion rate was zero and the deletion
rate was `r pars$rates[2]`.
```{r mevo-indels-caption}
ind_cap <- paste("Observed versus predicted insertions (left) and deletions (right)",
                 "for simulation of 2 variants along a phylogeny.")
```


I used the diagnostic console output from `jackalope` to examine how
each round of indel production compared to expectations.
For each "$\tau$-leap" (see Methods in main text for details), the console output 
the chromosome size, $\tau$, and the number of indels of each size that were produced.
This was important to do for each round because the chromosome size changed each time.
The predicted number of insertions of size $u$ was the product of the overall
insertion rate, $\tau$, chromosome size, and relative rate for size $u$.
The same goes for deletions.


```{r mevo-indels, fig.height=3, fig.width=6, fig.cap=ind_cap}
# Function to calculate the expected proportion of mutations that are
# insertions or deletions of a given size:
indel_pred <- function(u, j, b_len, chrom_size) {
    # < indel of type j and size u > / < indel of type j >
    M_ <- pars$M[j]
    a_ <- pars$a[j]
    u_ <- 1:M_
    all_ps <- {(u_ * M_) / (M_ - u_ + 1)}^(-a_)
    piju <- all_ps[u] / sum(all_ps)
    return(pars$rates[j] * piju * b_len * chrom_size)
}


# --------------------*
# Insertions ----
# --------------------*

# Only insertions:
insertions <- indels(rate = pars$rates[1],
                     max_length = pars$M[1],
                     a = pars$a[1])
ins_output <- capture.output(var_set <- create_variants(ref, vars_phylo(tree),
                                                        ins = insertions))

ins_info <- ins_output[grepl("^\\+-", ins_output)]
ins_info <- map(strsplit(ins_info, " \\| "), 
                ~ as.integer(strsplit(trimws(.x[2]), " ")[[1]]))
ins_info <- do.call(rbind, ins_info)

# branch length and chromosome size info
blcs_info <- ins_output[grepl("^\\+-", ins_output)]
blcs_info <- map(strsplit(blcs_info, " \\| "),
                 ~ as.numeric(strsplit(.x[1], " ")[[1]][2:3]))
blcs_info <- do.call(rbind, blcs_info)

ins_pred <- map(1:nrow(blcs_info),
                ~ indel_pred(1:pars$M[1], 1, blcs_info[.x,2], blcs_info[.x,1])) %>% 
  do.call(what = rbind)



ins_df <- tibble(pred = as.numeric(ins_pred),
                 obs = as.numeric(ins_info))



# --------------------*
# Deletions ----
# --------------------*

# Only deletions
deletions <- indels(rate = pars$rates[2],
                    max_length = pars$M[2],
                    a = pars$a[2])
del_output <- capture.output(var_set <- create_variants(ref, vars_phylo(tree),
                                                        del = deletions))

del_info <- del_output[grepl("^\\+-", del_output)]
del_info <- map(strsplit(del_info, " \\| "), 
                ~ as.integer(strsplit(trimws(.x[2]), " ")[[1]]))
del_info <- do.call(rbind, del_info)

# branch length and chromosome size info
blcs_info <- del_output[grepl("^\\+-", del_output)]
blcs_info <- map(strsplit(blcs_info, " \\| "),
                 ~ as.numeric(strsplit(.x[1], " ")[[1]][2:3]))
blcs_info <- do.call(rbind, blcs_info)


del_pred <- map(1:nrow(blcs_info),
                ~ indel_pred(1:pars$M[2], 2, blcs_info[.x,2], blcs_info[.x,1])) %>% 
  do.call(what = rbind)


del_df <- tibble(pred = as.numeric(del_pred),
                 obs = as.numeric(del_info))



# ---------*
# Plots
# ---------*


ins_p <- ins_df %>%
    ggplot(aes(pred, obs)) +
    geom_point(alpha = 0.5, shape = 1) +
    geom_abline(size = 1, slope = 1, intercept = 0, linetype = 2, color = "firebrick2") +
    scale_x_continuous("Predicted insertion count") +
    scale_y_continuous("Observed insertion count") +
    coord_equal()

del_p <- del_df %>%
    ggplot(aes(pred, obs)) +
    geom_point(alpha = 0.5, shape = 1) +
    geom_abline(size = 1, slope = 1, intercept = 0, linetype = 2, color = "firebrick2") +
    scale_x_continuous("Predicted deletion count") +
    scale_y_continuous("Observed deletion count") +
    coord_equal()



grid.newpage()
grid.draw(cbind(ggplotGrob(ins_p), ggplotGrob(del_p), size = "last"))

```






## Among-site variation

In this set of simulations, I included only substitutions because indels are not
affected by among-site variation.
I used the same substitution rate matrix as in the "Substitution rates" section.
For the discrete Gamma distribution used to create among-site variation, I used a shape
value of `r pars$shape` and split the distribution into 5 categories.


I used the diagnostic console output for this section.
Each time a substition was generated, the console output the position,
Gamma-distribution category (0 to 4 in this case), the original nucleotide, 
and the new nucleotide.
It also output when switching chromosomes, so I could match substitutions with
chromosomes.
I counted the number of mutations per chromosome, rate category, and type of substitution.
I compared these to predicted counts, which were the product of the
chromosome size, number of variants, and the transition-probability matrix for the
branch length of the phylogeny.



```{r mevo-site-var-cap}
sv_cap <- paste("Observed versus predicted number of mutations, based on each",
                "position's mutation rate modifier from a discrete Gamma distribution.",
                "The dashed line is the 1-to-1 line.")
gr_cap <- paste("Observed number of sites in each rate category.",
                "The red line is the predicted value for all categories.")
# Used to generate discrete Gamma values:
Rcpp::sourceCpp("supp/diagnostics-mevo.cpp")
```

```{r mevo-site-var, fig.cap = sv_cap}
# Only substitutions plus gammas
sub <- sub_TN93(alpha_1 = pars$alpha_1,
              alpha_2 = pars$alpha_2, beta = pars$beta,
              pi_tcag = pars$pi_tcag,
              gamma_shape = pars$shape, 
              gamma_k = 5)
gammas <- discrete_gamma(5, pars$shape)

# # Takes ~7 min
# svar_output <- capture.output(var_set <- create_variants(ref, vars_phylo(tree), sub = sub))
# saveRDS(svar_output, "supp/rds_files/svar_output.rds")
svar_output <- readRDS("supp/rds_files/svar_output.rds")


sub_info <- svar_output[grepl("^__", svar_output)]
sub_info <- map(strsplit(sub_info, " "), ~ .x[-1])

chroms <- gsub(">> ", "", svar_output[grepl("^>>", svar_output)])
chrom_starts <- which(grepl("^>>", svar_output))
chrom_subs <- map_int(1:length(chrom_starts),
        function(i) {
          end <- ifelse(i == length(chrom_starts), length(svar_output), 
                        chrom_starts[i+1]-1)
          n_subs <- sum(grepl("^__", svar_output[chrom_starts[i]:end]))
          return(n_subs)
        })

sub_df <- tibble(pos = map_int(sub_info, ~ as.integer(.x[[1]]) + 1L),
                 rate = map_int(sub_info, ~ as.integer(.x[[2]]) + 1L),
                 from = map_chr(sub_info, ~ substr(.x[3], 1, 1)),
                 to = map_chr(sub_info, ~ substr(.x[3], 3, 3))) %>% 
  mutate(chrom = rep(chroms, chrom_subs)) %>% 
  mutate_at(vars(chrom, rate, from, to), factor) %>% 
  group_by(chrom, rate, from, to, .drop = FALSE) %>% 
  summarize(obs = n()) %>% 
  ungroup() %>% 
  filter(from != to) %>% 
  mutate_at(vars(chrom, from, to), paste) %>% 
  mutate(rate = as.integer(paste(rate)))

e_muts <- map(1:5, function(i) {
  # mutations per base:
  Pt(max(ape::node.depth.edgelength(tree)), sub, i) * 
    # going from mutations/base to mutation/chromosome w/ rate i:
    matrix(ref$sizes()[1] / 5, 4, 4) *  
    # accounting for nucleotide frequencies in ref. genome:
    matrix(0.25, 4, 4) *
    # Accounting for the two variants:
    2 *
    # setting diagonal to zero:
    (1 - diag(4)) 
})


pred_df <- map_dfr(1:5, ~ rename(process_sub_mat(e_muts[[.x]], .x), 
                                 rate = id, pred = n)) %>% 
  mutate_at(vars(to, from), paste)


var_df <- left_join(sub_df, pred_df, by = c("to", "from", "rate"))


var_df %>%
    ggplot(aes(pred, obs)) +
    geom_point(shape = 1, alpha = 0.25) +
    geom_abline(size = 1, slope = 1, intercept = 0, linetype = 2, color = "firebrick2") +
    scale_x_continuous("Predicted mutation count") +
    scale_y_continuous("Observed mutation count") +
    coord_equal()
```



I also checked that the number of sites were approximately equal for each category.
This should be true because each rate category represents an equal area under
the probability density curve for the Gamma distribution.
The diagnostic output included when rates were created so that I could directly estimate
how many sites belonged in each rate category.



```{r mevo-site-var-rates, fig.cap = gr_cap}
rate_info <- svar_output[grepl("^~~", svar_output)]
rate_info <- sapply(strsplit(rate_info, " = "), function(x) trimws(x[2]))
rate_info <- lapply(strsplit(rate_info, " "), factor, levels = paste(0:4))
rate_mat <- do.call(rbind, lapply(rate_info, function(x) as.numeric(table(x))))

# This is the same for all:
gr_pred <- rowSums(rate_mat)[1] / ncol(rate_mat)

rate_df <- tibble(cat = rep(1:ncol(rate_mat), each = nrow(rate_mat)),
                  obs = as.numeric(rate_mat))

rate_df %>%
    ggplot(aes(obs)) +
    geom_histogram(bins = 20, fill = "dodgerblue", 
                   color = "gray70", size = 0.1) +
    geom_vline(xintercept = gr_pred, linetype = 2, color = "firebrick2") +
    xlab("Number in category") +
    ylab("Count")
```



## Phylogeny


For this section, I simulated 10 variants along the following phylogenetic tree:


```{r mevo-phylogeny-phylo, fig.height=4, fig.width=5, fig.cap="Phylogeny of 10 simulated variants."}
# Phylogenetic tree:
set.seed(89415648)
tree <- ape::rcoal(pars$n_vars)
tree$edge.length <- tree$edge.length * 0.1

mp_cap <- paste("Observed versus predicted numbers of unshared mutations",
                "between", pars$n_vars, "variants created by simulating along",
                "a phylogeny.",
                "Each point represents the number of unshared mutations between",
                "two variants for one of the", ref$n_chroms(), "chromosomes.",
                "The dashed line is the 1-to-1 line.")

plot(tree)
```


I used only substitutions in the simulations, to make matching mutations between
variants more straightforward.

I used the diagnostic console output for this section.
Each time a substition was generated, the console output the position,
Gamma-distribution category (always 1 in this case), the original nucleotide, 
and the new nucleotide.
It also output when switching to a new edge on the phylogeny, so I could
match substitutions with branch lengths.
I counted the number of mutations per tree edge and 
compared these to predicted counts, which were the product of the chromosome size
and the transition-probability matrix for the branch length.



```{r mevo-phylogeny-plot, fig.cap=mp_cap}
# Only substitutions again
sub <- sub_TN93(alpha_1 = pars$alpha_1,
              alpha_2 = pars$alpha_2, beta = pars$beta,
              pi_tcag = pars$pi_tcag)

# # Takes ~6.5 min
# phy_output <- capture.output(var_set <- create_variants(ref, vars_phylo(tree), sub = sub))
# saveRDS(phy_output, "supp/rds_files/phy_output.rds")
phy_output <- readRDS("supp/rds_files/phy_output.rds")

sub_info <- phy_output[grepl("^__", phy_output)]
sub_info <- map(strsplit(sub_info, " "), ~ .x[-1])

blens <- gsub("\\*\\* b_len ", "", phy_output[grepl("^\\*\\*", phy_output)])
blen_starts <- which(grepl("^\\*\\*", phy_output))
blen_obs <- map_int(1:length(blen_starts),
        function(i) {
          end <- ifelse(i == length(blen_starts), length(phy_output), 
                        blen_starts[i+1]-1)
          n_subs <- sum(grepl("^__", phy_output[blen_starts[i]:end]))
          return(n_subs)
        })

blen_pred <- blens %>%
    as.numeric() %>% 
    map_dbl(function(t) {
        pred_mat <- Pt(t, .sub = sub) * 
            matrix(ref$sizes()[1], 4, 4) * 
            matrix(0.25, 4, 4) * 
            (1 - diag(4))
        return(sum(pred_mat))
    })

phy_df <- tibble(obs = blen_obs, pred = blen_pred)


phy_df %>%
    ggplot(aes(pred, obs)) +
    geom_point(alpha = 0.5, shape = 1) +
    geom_abline(size = 1, slope = 1, intercept = 0, linetype = 2, color = "firebrick2") +
    scale_x_continuous("Predicted mutations count") +
    scale_y_continuous("Observed mutations count") +
    coord_equal() +
    NULL

```
